{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/zhenhao/miniconda3/lib/python3.8/site-packages/pysradb/utils.py:14: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from tqdm.autonotebook import tqdm\n"
     ]
    }
   ],
   "source": [
    "from pysradb import SRAweb\n",
    "\n",
    "db = SRAweb()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_814/4160885475.py:3: DtypeWarning: Columns (61,63,65,74,82,83) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gtdb_metadata = pd.read_csv('/mnt/c/Users/zhenh/tax_data/bac120_metadata.tsv',sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "gtdb_metadata = pd.read_csv('/mnt/c/Users/zhenh/tax_data/bac120_metadata.tsv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['accession', 'ambiguous_bases', 'checkm_completeness',\n",
       "       'checkm_contamination', 'checkm_marker_count', 'checkm_marker_lineage',\n",
       "       'checkm_marker_set_count', 'checkm_strain_heterogeneity',\n",
       "       'coding_bases', 'coding_density',\n",
       "       ...\n",
       "       'ssu_silva_blast_align_len', 'ssu_silva_blast_bitscore',\n",
       "       'ssu_silva_blast_evalue', 'ssu_silva_blast_perc_identity',\n",
       "       'ssu_silva_blast_subject_id', 'ssu_silva_taxonomy', 'total_gap_length',\n",
       "       'trna_aa_count', 'trna_count', 'trna_selenocysteine_count'],\n",
       "      dtype='object', length=110)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_metadata.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ncbi_organism_name Streptococcus pneumoniae\n"
     ]
    }
   ],
   "source": [
    "for i, a in enumerate(gtdb_metadata.loc[0]):\n",
    "    #print(gtdb_metadata.columns[i], a)\n",
    "    #print(a)\n",
    "    if a == \"Streptococcus pneumoniae\":\n",
    "        print(gtdb_metadata.columns[i], a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accession\n",
      "ambiguous_bases\n",
      "checkm_completeness\n",
      "checkm_contamination\n",
      "checkm_marker_count\n",
      "checkm_marker_lineage\n",
      "checkm_marker_set_count\n",
      "checkm_strain_heterogeneity\n",
      "coding_bases\n",
      "coding_density\n",
      "contig_count\n",
      "gc_count\n",
      "gc_percentage\n",
      "genome_size\n",
      "gtdb_genome_representative\n",
      "gtdb_representative\n",
      "gtdb_taxonomy\n",
      "gtdb_type_designation_ncbi_taxa\n",
      "gtdb_type_designation_ncbi_taxa_sources\n",
      "gtdb_type_species_of_genus\n",
      "l50_contigs\n",
      "l50_scaffolds\n",
      "longest_contig\n",
      "longest_scaffold\n",
      "lsu_23s_contig_len\n",
      "lsu_23s_count\n",
      "lsu_23s_length\n",
      "lsu_23s_query_id\n",
      "lsu_5s_contig_len\n",
      "lsu_5s_count\n",
      "lsu_5s_length\n",
      "lsu_5s_query_id\n",
      "lsu_silva_23s_blast_align_len\n",
      "lsu_silva_23s_blast_bitscore\n",
      "lsu_silva_23s_blast_evalue\n",
      "lsu_silva_23s_blast_perc_identity\n",
      "lsu_silva_23s_blast_subject_id\n",
      "lsu_silva_23s_taxonomy\n",
      "mean_contig_length\n",
      "mean_scaffold_length\n",
      "mimag_high_quality\n",
      "mimag_low_quality\n",
      "mimag_medium_quality\n",
      "n50_contigs\n",
      "n50_scaffolds\n",
      "ncbi_assembly_level\n",
      "ncbi_assembly_name\n",
      "ncbi_assembly_type\n",
      "ncbi_bioproject\n",
      "ncbi_biosample\n",
      "ncbi_contig_count\n",
      "ncbi_contig_n50\n",
      "ncbi_country\n",
      "ncbi_date\n",
      "ncbi_genbank_assembly_accession\n",
      "ncbi_genome_category\n",
      "ncbi_genome_representation\n",
      "ncbi_isolate\n",
      "ncbi_isolation_source\n",
      "ncbi_lat_lon\n",
      "ncbi_molecule_count\n",
      "ncbi_ncrna_count\n",
      "ncbi_organism_name\n",
      "ncbi_protein_count\n",
      "ncbi_refseq_category\n",
      "ncbi_rrna_count\n",
      "ncbi_scaffold_count\n",
      "ncbi_scaffold_l50\n",
      "ncbi_scaffold_n50\n",
      "ncbi_scaffold_n75\n",
      "ncbi_scaffold_n90\n",
      "ncbi_seq_rel_date\n",
      "ncbi_spanned_gaps\n",
      "ncbi_species_taxid\n",
      "ncbi_ssu_count\n",
      "ncbi_strain_identifiers\n",
      "ncbi_submitter\n",
      "ncbi_taxid\n",
      "ncbi_taxonomy\n",
      "ncbi_taxonomy_unfiltered\n",
      "ncbi_total_gap_length\n",
      "ncbi_total_length\n",
      "ncbi_translation_table\n",
      "ncbi_trna_count\n",
      "ncbi_type_material_designation\n",
      "ncbi_ungapped_length\n",
      "ncbi_unspanned_gaps\n",
      "ncbi_wgs_master\n",
      "protein_count\n",
      "scaffold_count\n",
      "ssu_contig_len\n",
      "ssu_count\n",
      "ssu_gg_blast_align_len\n",
      "ssu_gg_blast_bitscore\n",
      "ssu_gg_blast_evalue\n",
      "ssu_gg_blast_perc_identity\n",
      "ssu_gg_blast_subject_id\n",
      "ssu_gg_taxonomy\n",
      "ssu_length\n",
      "ssu_query_id\n",
      "ssu_silva_blast_align_len\n",
      "ssu_silva_blast_bitscore\n",
      "ssu_silva_blast_evalue\n",
      "ssu_silva_blast_perc_identity\n",
      "ssu_silva_blast_subject_id\n",
      "ssu_silva_taxonomy\n",
      "total_gap_length\n",
      "trna_aa_count\n",
      "trna_count\n",
      "trna_selenocysteine_count\n"
     ]
    }
   ],
   "source": [
    "for feature in gtdb_metadata.columns:\n",
    "    print(feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbi_species_taxid</th>\n",
       "      <th>ncbi_taxid</th>\n",
       "      <th>ncbi_organism_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1313</td>\n",
       "      <td>1313</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1313</td>\n",
       "      <td>1313</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216816</td>\n",
       "      <td>1457183</td>\n",
       "      <td>Bifidobacterium longum subsp. infantis EK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>661</td>\n",
       "      <td>661</td>\n",
       "      <td>Photobacterium angustum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1359</td>\n",
       "      <td>1295826</td>\n",
       "      <td>Lactococcus lactis subsp. cremoris KW2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ncbi_species_taxid  ncbi_taxid                          ncbi_organism_name\n",
       "0                1313        1313                    Streptococcus pneumoniae\n",
       "1                1313        1313                    Streptococcus pneumoniae\n",
       "2              216816     1457183  Bifidobacterium longum subsp. infantis EK3\n",
       "3                 661         661                     Photobacterium angustum\n",
       "4                1359     1295826      Lactococcus lactis subsp. cremoris KW2"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_metadata.head()[[\"ncbi_species_taxid\", \"ncbi_taxid\", \"ncbi_organism_name\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "57939"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_set = set(gtdb_metadata[\"ncbi_species_taxid\"])\n",
    "len(species_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtdb_metadata.loc[0][\"ncbi_organism_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[22], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43mdb\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msearch_sra\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgtdb_metadata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mncbi_organism_name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pysradb/sradb.py:1172\u001b[0m, in \u001b[0;36mSRAdb.search_sra\u001b[0;34m(self, search_str, out_type, assay, sample_attribute, detailed, expand_sample_attributes, output_read_lengths)\u001b[0m\n\u001b[1;32m   1142\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msearch_sra\u001b[39m(\n\u001b[1;32m   1143\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1144\u001b[0m     search_str,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1155\u001b[0m     output_read_lengths\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m   1156\u001b[0m ):\n\u001b[1;32m   1157\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Search SRA for any search term.\u001b[39;00m\n\u001b[1;32m   1158\u001b[0m \n\u001b[1;32m   1159\u001b[0m \u001b[38;5;124;03m    Parameters\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1170\u001b[0m \u001b[38;5;124;03m              Dataframe with relevant query results\u001b[39;00m\n\u001b[1;32m   1171\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1172\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msra_metadata\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1173\u001b[0m \u001b[43m        \u001b[49m\u001b[43msearch_str\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1174\u001b[0m \u001b[43m        \u001b[49m\u001b[43mout_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mout_type\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1175\u001b[0m \u001b[43m        \u001b[49m\u001b[43massay\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43massay\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1176\u001b[0m \u001b[43m        \u001b[49m\u001b[43msample_attribute\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msample_attribute\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1177\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdetailed\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdetailed\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1178\u001b[0m \u001b[43m        \u001b[49m\u001b[43mexpand_sample_attributes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexpand_sample_attributes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1179\u001b[0m \u001b[43m        \u001b[49m\u001b[43moutput_read_lengths\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_read_lengths\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1180\u001b[0m \u001b[43m        \u001b[49m\u001b[43macc_is_searchstr\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m   1181\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pysradb/sraweb.py:415\u001b[0m, in \u001b[0;36mSRAweb.sra_metadata\u001b[0;34m(self, srp, sample_attribute, detailed, expand_sample_attributes, output_read_lengths, **kwargs)\u001b[0m\n\u001b[1;32m    406\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msra_metadata\u001b[39m(\n\u001b[1;32m    407\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    408\u001b[0m     srp,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    413\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs\n\u001b[1;32m    414\u001b[0m ):\n\u001b[0;32m--> 415\u001b[0m     esummary_result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_esummary_response\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msra\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msrp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    416\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    417\u001b[0m         uids \u001b[38;5;241m=\u001b[39m esummary_result[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muids\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pysradb/sraweb.py:298\u001b[0m, in \u001b[0;36mSRAweb.get_esummary_response\u001b[0;34m(self, db, term, usehistory)\u001b[0m\n\u001b[1;32m    296\u001b[0m payload \u001b[38;5;241m=\u001b[39m OrderedDict(payload)\n\u001b[1;32m    297\u001b[0m payload[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mretstart\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m retstart\n\u001b[0;32m--> 298\u001b[0m request \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    299\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbase_url\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mesummary\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mOrderedDict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpayload\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    300\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    301\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    302\u001b[0m     response \u001b[38;5;241m=\u001b[39m request\u001b[38;5;241m.\u001b[39mjson()\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/requests/sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    706\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/requests/adapters.py:486\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    483\u001b[0m     timeout \u001b[38;5;241m=\u001b[39m TimeoutSauce(connect\u001b[38;5;241m=\u001b[39mtimeout, read\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    485\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 486\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    487\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    488\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    489\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    490\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    491\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    492\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    495\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    496\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    497\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    498\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    500\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ProtocolError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    501\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(err, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:715\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    712\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_prepare_proxy(conn)\n\u001b[1;32m    714\u001b[0m \u001b[38;5;66;03m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 715\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    716\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    717\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    718\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    719\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    720\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    721\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    722\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    723\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    725\u001b[0m \u001b[38;5;66;03m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    726\u001b[0m \u001b[38;5;66;03m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    727\u001b[0m \u001b[38;5;66;03m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    728\u001b[0m \u001b[38;5;66;03m# mess.\u001b[39;00m\n\u001b[1;32m    729\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:404\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    402\u001b[0m \u001b[38;5;66;03m# Trigger any extra validation we need to do.\u001b[39;00m\n\u001b[1;32m    403\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 404\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    405\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    406\u001b[0m     \u001b[38;5;66;03m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[39;00m\n\u001b[1;32m    407\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mconn\u001b[38;5;241m.\u001b[39mtimeout)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:1058\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   1056\u001b[0m \u001b[38;5;66;03m# Force connect early to allow us to validate the connection.\u001b[39;00m\n\u001b[1;32m   1057\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(conn, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msock\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):  \u001b[38;5;66;03m# AppEngine might not have  `.sock`\u001b[39;00m\n\u001b[0;32m-> 1058\u001b[0m     \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1060\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_verified:\n\u001b[1;32m   1061\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m   1062\u001b[0m         (\n\u001b[1;32m   1063\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnverified HTTPS request is being made to host \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1068\u001b[0m         InsecureRequestWarning,\n\u001b[1;32m   1069\u001b[0m     )\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/connection.py:363\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    361\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mconnect\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    362\u001b[0m     \u001b[38;5;66;03m# Add certificate verification\u001b[39;00m\n\u001b[0;32m--> 363\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_new_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    364\u001b[0m     hostname \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost\n\u001b[1;32m    365\u001b[0m     tls_in_tls \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/connection.py:174\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    171\u001b[0m     extra_kw[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msocket_options\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msocket_options\n\u001b[1;32m    173\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 174\u001b[0m     conn \u001b[38;5;241m=\u001b[39m \u001b[43mconnection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_connection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    175\u001b[0m \u001b[43m        \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_dns_host\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mport\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mextra_kw\u001b[49m\n\u001b[1;32m    176\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    178\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m SocketTimeout:\n\u001b[1;32m    179\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m ConnectTimeoutError(\n\u001b[1;32m    180\u001b[0m         \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    181\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mConnection to \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m timed out. (connect timeout=\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    182\u001b[0m         \u001b[38;5;241m%\u001b[39m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtimeout),\n\u001b[1;32m    183\u001b[0m     )\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/util/connection.py:85\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address, socket_options)\u001b[0m\n\u001b[1;32m     83\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m source_address:\n\u001b[1;32m     84\u001b[0m         sock\u001b[38;5;241m.\u001b[39mbind(source_address)\n\u001b[0;32m---> 85\u001b[0m     \u001b[43msock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43msa\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     86\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m sock\n\u001b[1;32m     88\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m socket\u001b[38;5;241m.\u001b[39merror \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "res = db.search_sra(gtdb_metadata.loc[0][\"ncbi_organism_name\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>study_title</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>experiment_desc</th>\n",
       "      <th>organism_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>...</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>type-material</th>\n",
       "      <th>biosamplemodel</th>\n",
       "      <th>sub_species</th>\n",
       "      <th>ena_fastq_http</th>\n",
       "      <th>ena_fastq_http_1</th>\n",
       "      <th>ena_fastq_http_2</th>\n",
       "      <th>ena_fastq_ftp</th>\n",
       "      <th>ena_fastq_ftp_1</th>\n",
       "      <th>ena_fastq_ftp_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR24651172</td>\n",
       "      <td>SRP438478</td>\n",
       "      <td>Genome sequencing and hybrid assembly of micro...</td>\n",
       "      <td>SRX20431359</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>537012</td>\n",
       "      <td>Bacteroides cellulosilyticus DSM 14838</td>\n",
       "      <td>Nanopore-sequencing-of-Bacteroides-cellulosily...</td>\n",
       "      <td>WGS</td>\n",
       "      <td>...</td>\n",
       "      <td>not collected</td>\n",
       "      <td>type strain of Bacteroides cellulosilyticus</td>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR24651174</td>\n",
       "      <td>SRP438478</td>\n",
       "      <td>Genome sequencing and hybrid assembly of micro...</td>\n",
       "      <td>SRX20431357</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>84112</td>\n",
       "      <td>Eggerthella lenta</td>\n",
       "      <td>Nanopore-sequencing-of-Eggerthella-lenta-DSM-2243</td>\n",
       "      <td>WGS</td>\n",
       "      <td>...</td>\n",
       "      <td>not collected</td>\n",
       "      <td>type strain of Eggerthella lenta</td>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR24651176</td>\n",
       "      <td>SRP438478</td>\n",
       "      <td>Genome sequencing and hybrid assembly of micro...</td>\n",
       "      <td>SRX20431355</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>338188</td>\n",
       "      <td>Bacteroides finegoldii</td>\n",
       "      <td>Nanopore-sequencing-of-Bacteroides-finegoldii-...</td>\n",
       "      <td>WGS</td>\n",
       "      <td>...</td>\n",
       "      <td>not collected</td>\n",
       "      <td>type strain of Bacteroides finegoldii</td>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR24651177</td>\n",
       "      <td>SRP438478</td>\n",
       "      <td>Genome sequencing and hybrid assembly of micro...</td>\n",
       "      <td>SRX20431354</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>435591</td>\n",
       "      <td>Parabacteroides distasonis ATCC 8503</td>\n",
       "      <td>Nanopore-sequencing-of-Parabacteroides-distaso...</td>\n",
       "      <td>WGS</td>\n",
       "      <td>...</td>\n",
       "      <td>not collected</td>\n",
       "      <td>type strain of Parabacteroides distasonis</td>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR24651179</td>\n",
       "      <td>SRP438478</td>\n",
       "      <td>Genome sequencing and hybrid assembly of micro...</td>\n",
       "      <td>SRX20431352</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>84030</td>\n",
       "      <td>Lacrimispora saccharolytica</td>\n",
       "      <td>Nanopore-sequencing-of-Lacrimispora-saccharoly...</td>\n",
       "      <td>WGS</td>\n",
       "      <td>...</td>\n",
       "      <td>not collected</td>\n",
       "      <td>type strain of Clostridium saccharolyticum</td>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession study_accession  \\\n",
       "1   SRR24651172       SRP438478   \n",
       "3   SRR24651174       SRP438478   \n",
       "5   SRR24651176       SRP438478   \n",
       "6   SRR24651177       SRP438478   \n",
       "8   SRR24651179       SRP438478   \n",
       "\n",
       "                                         study_title experiment_accession  \\\n",
       "1  Genome sequencing and hybrid assembly of micro...          SRX20431359   \n",
       "3  Genome sequencing and hybrid assembly of micro...          SRX20431357   \n",
       "5  Genome sequencing and hybrid assembly of micro...          SRX20431355   \n",
       "6  Genome sequencing and hybrid assembly of micro...          SRX20431354   \n",
       "8  Genome sequencing and hybrid assembly of micro...          SRX20431352   \n",
       "\n",
       "                                    experiment_title  \\\n",
       "1  PCR free Oxford Nanopore sequencing of genomic...   \n",
       "3  PCR free Oxford Nanopore sequencing of genomic...   \n",
       "5  PCR free Oxford Nanopore sequencing of genomic...   \n",
       "6  PCR free Oxford Nanopore sequencing of genomic...   \n",
       "8  PCR free Oxford Nanopore sequencing of genomic...   \n",
       "\n",
       "                                     experiment_desc organism_taxid  \\\n",
       "1  PCR free Oxford Nanopore sequencing of genomic...         537012   \n",
       "3  PCR free Oxford Nanopore sequencing of genomic...          84112   \n",
       "5  PCR free Oxford Nanopore sequencing of genomic...         338188   \n",
       "6  PCR free Oxford Nanopore sequencing of genomic...         435591   \n",
       "8  PCR free Oxford Nanopore sequencing of genomic...          84030   \n",
       "\n",
       "                            organism_name  \\\n",
       "1  Bacteroides cellulosilyticus DSM 14838   \n",
       "3                       Eggerthella lenta   \n",
       "5                  Bacteroides finegoldii   \n",
       "6    Parabacteroides distasonis ATCC 8503   \n",
       "8             Lacrimispora saccharolytica   \n",
       "\n",
       "                                        library_name library_strategy  ...  \\\n",
       "1  Nanopore-sequencing-of-Bacteroides-cellulosily...              WGS  ...   \n",
       "3  Nanopore-sequencing-of-Eggerthella-lenta-DSM-2243              WGS  ...   \n",
       "5  Nanopore-sequencing-of-Bacteroides-finegoldii-...              WGS  ...   \n",
       "6  Nanopore-sequencing-of-Parabacteroides-distaso...              WGS  ...   \n",
       "8  Nanopore-sequencing-of-Lacrimispora-saccharoly...              WGS  ...   \n",
       "\n",
       "  ref_biomaterial                                type-material  \\\n",
       "1   not collected  type strain of Bacteroides cellulosilyticus   \n",
       "3   not collected             type strain of Eggerthella lenta   \n",
       "5   not collected        type strain of Bacteroides finegoldii   \n",
       "6   not collected    type strain of Parabacteroides distasonis   \n",
       "8   not collected   type strain of Clostridium saccharolyticum   \n",
       "\n",
       "                biosamplemodel sub_species ena_fastq_http ena_fastq_http_1  \\\n",
       "1  MIGS/MIMS/MIMARKS.human-gut        <NA>           <NA>             <NA>   \n",
       "3  MIGS/MIMS/MIMARKS.human-gut        <NA>           <NA>             <NA>   \n",
       "5  MIGS/MIMS/MIMARKS.human-gut        <NA>           <NA>             <NA>   \n",
       "6  MIGS/MIMS/MIMARKS.human-gut        <NA>           <NA>             <NA>   \n",
       "8  MIGS/MIMS/MIMARKS.human-gut        <NA>           <NA>             <NA>   \n",
       "\n",
       "  ena_fastq_http_2 ena_fastq_ftp ena_fastq_ftp_1 ena_fastq_ftp_2  \n",
       "1             <NA>          <NA>            <NA>            <NA>  \n",
       "3             <NA>          <NA>            <NA>            <NA>  \n",
       "5             <NA>          <NA>            <NA>            <NA>  \n",
       "6             <NA>          <NA>            <NA>            <NA>  \n",
       "8             <NA>          <NA>            <NA>            <NA>  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_df = db.sra_metadata(\"SRP438478\", detailed=True)\n",
    "\n",
    "study_df = study_df[study_df[\"instrument_model_desc\"] == \"OXFORD_NANOPORE\"]\n",
    "study_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR24651172\n",
      "SRP438478\n",
      "Genome sequencing and hybrid assembly of microbial isolates derived from the human gut.\n",
      "SRX20431359\n",
      "PCR free Oxford Nanopore sequencing of genomic DNA from Bacteroides-cellulosilyticus-DSM-14838\n",
      "PCR free Oxford Nanopore sequencing of genomic DNA from Bacteroides-cellulosilyticus-DSM-14838\n",
      "537012\n",
      "Bacteroides cellulosilyticus DSM 14838\n",
      "Nanopore-sequencing-of-Bacteroides-cellulosilyticus-DSM-14838\n",
      "WGS\n",
      "GENOMIC\n",
      "RANDOM\n",
      "SINGLE\n",
      "SRS17736875\n",
      "<NA>\n",
      "SAMN34894971\n",
      "PRJNA885826\n",
      "MinION\n",
      "MinION\n",
      "OXFORD_NANOPORE\n",
      "178241\n",
      "2652210907\n",
      "178241\n",
      "2938492130\n",
      "Bacteroides-cellulosilyticus-DSM-14838__barcode15.fastq.gz\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR24651172/SRR24651172\n",
      "worldwide\n",
      "anonymous\n",
      "<NA>\n",
      "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos3/sra-pub-run-36/SRR024/24651/SRR24651172/SRR24651172.1\n",
      "worldwide\n",
      "anonymous\n",
      "gs://sra-pub-run-13/SRR24651172/SRR24651172.1\n",
      "gs.US\n",
      "gcp identity\n",
      "<NA>\n",
      "DSM 14838\n",
      "not collected\n",
      "not applicable\n",
      "not applicable\n",
      "not applicable\n",
      "not applicable\n",
      "Homo sapiens\n",
      "not collected\n",
      "not collected\n",
      "not collected\n",
      "not collected\n",
      "type strain of Bacteroides cellulosilyticus\n",
      "MIGS/MIMS/MIMARKS.human-gut\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n"
     ]
    }
   ],
   "source": [
    "for feature in study_df.columns:\n",
    "    print(study_df[feature].iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['run_accession', 'study_accession', 'study_title',\n",
       "       'experiment_accession', 'experiment_title', 'experiment_desc',\n",
       "       'organism_taxid', 'organism_name', 'library_name', 'library_strategy',\n",
       "       'library_source', 'library_selection', 'library_layout',\n",
       "       'sample_accession', 'sample_title', 'biosample', 'bioproject',\n",
       "       'instrument', 'instrument_model', 'instrument_model_desc',\n",
       "       'total_spots', 'total_size', 'run_total_spots', 'run_total_bases',\n",
       "       'run_alias', 'public_filename', 'public_size', 'public_date',\n",
       "       'public_md5', 'public_version', 'public_semantic_name',\n",
       "       'public_supertype', 'public_sratoolkit', 'aws_url', 'aws_free_egress',\n",
       "       'aws_access_type', 'public_url', 'ncbi_url', 'ncbi_free_egress',\n",
       "       'ncbi_access_type', 'gcp_url', 'gcp_free_egress', 'gcp_access_type',\n",
       "       'experiment_alias', 'strain', 'collection_date', 'env_broad_scale',\n",
       "       'env_local_scale', 'env_medium', 'geo_loc_name', 'host',\n",
       "       'isol_growth_condt', 'lat_lon', 'num_replicons', 'ref_biomaterial',\n",
       "       'type-material', 'biosamplemodel', 'sub_species', 'ena_fastq_http',\n",
       "       'ena_fastq_http_1', 'ena_fastq_http_2', 'ena_fastq_ftp',\n",
       "       'ena_fastq_ftp_1', 'ena_fastq_ftp_2'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR24651171 SRR24651172 SRR24651173 SRR24651174 SRR24651175 SRR24651176 SRR24651177 SRR24651178 SRR24651179 SRR24651180 SRR24651181 SRR24651182 SRR24651183 SRR24651184 SRR24651185 SRR24651186 SRR24651187 SRR24651188 SRR24651189 SRR24651190 SRR24651191 SRR24651192 SRR24651193 SRR24651194 SRR24651195 SRR24651196 SRR24651197 SRR24651198 SRR24651199 SRR24651200 SRR24651201 SRR24651202 SRR24651203 SRR24651204 SRR24651205 SRR24651206 SRR24651207 SRR24651208 SRR24651209 SRR24651210 SRR24651211 SRR24651212 SRR24651213 SRR24651214 SRR24651215 SRR24651216 SRR24651217 SRR24651218 SRR24651219 SRR24651220 SRR24651221 SRR24651222 SRR24651223 SRR24651224 SRR24651225 SRR24651226 SRR24651227 SRR24651228 SRR24651229 SRR24651230 SRR24651231 SRR24651232 SRR24651233 SRR24651234 SRR24651235 SRR24651236 SRR24651237 SRR24651238 SRR24651239 SRR24651240 SRR24651241 SRR24651242 SRR24651243 SRR24651244 SRR24651245 SRR24651246 SRR24651247 SRR24651248 SRR24651249 SRR24651250 SRR24651251 SRR24651252 SRR24651253 SRR24651254 SRR24651255 SRR24651256 SRR24651257 SRR24651258 SRR24651259 SRR24651260 SRR24651261 SRR24651262 SRR24651263 SRR24651264 SRR24651265 SRR24651266 SRR24651267 SRR24651268 SRR24651269 SRR24651270 SRR24651271 SRR24651272 SRR24651273 SRR24651274 SRR24651275 SRR24651276 SRR24651277 SRR24651278 SRR24651279 SRR24651280 SRR24651281 SRR24651282 SRR24651283 SRR24651284 SRR24651285 SRR24651286 SRR24651287 SRR24651288 SRR24651289 SRR24651290 SRR24651291 SRR24651292 SRR24651293 SRR24651294 SRR24651295 SRR24651296 SRR24651297 SRR24651298 SRR24651299 SRR24651300 SRR24651301 SRR24651302 SRR24651303 SRR24651304 SRR24651305 SRR24651306 SRR24651307 SRR24651308 SRR24651309 SRR24651310 SRR24651311 SRR24651312 SRR24651313 SRR24651314 SRR24651315 SRR24651316 SRR24651317 SRR24651318 SRR24651319 SRR24651320 SRR24651321 SRR24651322 SRR24651323 SRR24651324 SRR24651325 SRR24651326 SRR24651327 SRR24651328 "
     ]
    }
   ],
   "source": [
    "for accession in study_df[\"run_accession\"]:\n",
    "    print(accession, end=\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
