{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/zhenhao/.local/lib/python3.10/site-packages/pysradb/utils.py:14: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from tqdm.autonotebook import tqdm\n"
     ]
    }
   ],
   "source": [
    "from pysradb import SRAweb\n",
    "\n",
    "db = SRAweb()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pysradb.search import EnaSearch, SraSearch\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1444/3778821307.py:3: DtypeWarning: Columns (61,63,65,74,82,83) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gtdb_metadata = pd.read_csv('./bac120_metadata.tsv',sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "gtdb_metadata = pd.read_csv('./bac120_metadata.tsv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbi_taxid</th>\n",
       "      <th>ncbi_organism_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17950</th>\n",
       "      <td>9</td>\n",
       "      <td>Buchnera aphidicola</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244877</th>\n",
       "      <td>14</td>\n",
       "      <td>Dictyoglomus thermophilum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183713</th>\n",
       "      <td>23</td>\n",
       "      <td>Shewanella colwelliana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378860</th>\n",
       "      <td>24</td>\n",
       "      <td>Shewanella putrefaciens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146850</th>\n",
       "      <td>25</td>\n",
       "      <td>Shewanella hanedai</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95168</th>\n",
       "      <td>2976476</td>\n",
       "      <td>Chryseobacterium sp. pc1-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386035</th>\n",
       "      <td>2976537</td>\n",
       "      <td>Citrobacter sp. XY323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193233</th>\n",
       "      <td>2976703</td>\n",
       "      <td>Mesorhizobium sp. P13.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386552</th>\n",
       "      <td>2976705</td>\n",
       "      <td>Pantoea sp. XY16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205319</th>\n",
       "      <td>2978683</td>\n",
       "      <td>Burkholderia cenocepacia MC0-3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82416 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ncbi_taxid              ncbi_organism_name\n",
       "17950            9             Buchnera aphidicola\n",
       "244877          14       Dictyoglomus thermophilum\n",
       "183713          23          Shewanella colwelliana\n",
       "378860          24         Shewanella putrefaciens\n",
       "146850          25              Shewanella hanedai\n",
       "...            ...                             ...\n",
       "95168      2976476     Chryseobacterium sp. pc1-10\n",
       "386035     2976537           Citrobacter sp. XY323\n",
       "193233     2976703         Mesorhizobium sp. P13.3\n",
       "386552     2976705                Pantoea sp. XY16\n",
       "205319     2978683  Burkholderia cenocepacia MC0-3\n",
       "\n",
       "[82416 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_df = gtdb_metadata[[\"ncbi_taxid\", \"ncbi_organism_name\"]]\n",
    "species_df = species_df.groupby(\"ncbi_taxid\").sample(1)\n",
    "species_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "species_df.iloc[1]['ncbi_taxid']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 626/626 [00:29<00:00, 20.88it/s]\n"
     ]
    }
   ],
   "source": [
    "instance = SraSearch(2, 10000, query=\"R10.4.1\", source=\"genomic\", strategy=\"wgs\", platform=\"oxford nanopore\")\n",
    "instance.search()\n",
    "res_df = instance.get_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>sample_taxon_id</th>\n",
       "      <th>sample_scientific_name</th>\n",
       "      <th>experiment_library_strategy</th>\n",
       "      <th>experiment_library_source</th>\n",
       "      <th>experiment_library_selection</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>experiment_instrument_model</th>\n",
       "      <th>pool_member_spots</th>\n",
       "      <th>run_1_size</th>\n",
       "      <th>run_1_accession</th>\n",
       "      <th>run_1_total_spots</th>\n",
       "      <th>run_1_total_bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRP500767</td>\n",
       "      <td>SRX24214614</td>\n",
       "      <td>ONT seq of Pf_T2</td>\n",
       "      <td>463681</td>\n",
       "      <td>Penicillium fuscoglaucum</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>SRS20984821</td>\n",
       "      <td>Pf_T2</td>\n",
       "      <td>MinION</td>\n",
       "      <td>1723079</td>\n",
       "      <td>13484052567</td>\n",
       "      <td>SRR28616354</td>\n",
       "      <td>1723079</td>\n",
       "      <td>15302485972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRP499270</td>\n",
       "      <td>SRX24128005</td>\n",
       "      <td>Basecalled ONT Sequences of Spirobacillus cien...</td>\n",
       "      <td>495820</td>\n",
       "      <td>Spirobacillus cienkowskii</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS20913379</td>\n",
       "      <td>Spirobacillus_cienkowskii_FI-SP1</td>\n",
       "      <td>MinION</td>\n",
       "      <td>5283013</td>\n",
       "      <td>3913249236</td>\n",
       "      <td>SRR28527199</td>\n",
       "      <td>5283013</td>\n",
       "      <td>4450905037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRP496639</td>\n",
       "      <td>SRX24019922</td>\n",
       "      <td>ONT sequencing of Staphylococcus aureus TUM20914</td>\n",
       "      <td>1280</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS20798084</td>\n",
       "      <td>Staphylococcus aureus TUM20914</td>\n",
       "      <td>MinION</td>\n",
       "      <td>69084</td>\n",
       "      <td>340431788</td>\n",
       "      <td>SRR28415513</td>\n",
       "      <td>69084</td>\n",
       "      <td>368340750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRP496639</td>\n",
       "      <td>SRX24019921</td>\n",
       "      <td>ONT sequencing of Staphylococcus aureus TUM20913</td>\n",
       "      <td>1280</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS20798083</td>\n",
       "      <td>Staphylococcus aureus TUM20913</td>\n",
       "      <td>MinION</td>\n",
       "      <td>184459</td>\n",
       "      <td>681816343</td>\n",
       "      <td>SRR28415514</td>\n",
       "      <td>184459</td>\n",
       "      <td>735944460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRP496639</td>\n",
       "      <td>SRX24019920</td>\n",
       "      <td>ONT sequencing of Staphylococcus aureus TUM20888</td>\n",
       "      <td>1280</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS20798082</td>\n",
       "      <td>Staphylococcus aureus TUM20888</td>\n",
       "      <td>MinION</td>\n",
       "      <td>209063</td>\n",
       "      <td>1235881070</td>\n",
       "      <td>SRR28415515</td>\n",
       "      <td>209063</td>\n",
       "      <td>1349673232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>SRP417920</td>\n",
       "      <td>SRX19052651</td>\n",
       "      <td>Whole Genome Sequencing of three Canadian stra...</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS16469023</td>\n",
       "      <td>HH45</td>\n",
       "      <td>MinION</td>\n",
       "      <td>251505</td>\n",
       "      <td>376551024</td>\n",
       "      <td>SRR23100674</td>\n",
       "      <td>251505</td>\n",
       "      <td>437387806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>SRP413991</td>\n",
       "      <td>SRX18782374</td>\n",
       "      <td>Oxford Nanopore Q20 400bps sequencing of adult...</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS16216412</td>\n",
       "      <td>dmel-bdgp-iso1-Q20</td>\n",
       "      <td>MinION</td>\n",
       "      <td>1937835</td>\n",
       "      <td>18615677517</td>\n",
       "      <td>SRR22822929</td>\n",
       "      <td>1937835</td>\n",
       "      <td>21140714134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>SRP413991</td>\n",
       "      <td>SRX18782373</td>\n",
       "      <td>Oxford Nanopore Q20 260bps sequencing of adult...</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS16216412</td>\n",
       "      <td>dmel-bdgp-iso1-Q20</td>\n",
       "      <td>MinION</td>\n",
       "      <td>1481121</td>\n",
       "      <td>16566954959</td>\n",
       "      <td>SRR22822930</td>\n",
       "      <td>1481121</td>\n",
       "      <td>19204298459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>SRP413936</td>\n",
       "      <td>SRX18776194</td>\n",
       "      <td>Madurella fahalii CBS129176</td>\n",
       "      <td>1157608</td>\n",
       "      <td>Madurella fahalii</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS16210388</td>\n",
       "      <td>CBS129176</td>\n",
       "      <td>MinION</td>\n",
       "      <td>696330</td>\n",
       "      <td>2085105473</td>\n",
       "      <td>SRR22816638</td>\n",
       "      <td>696330</td>\n",
       "      <td>2327283485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>SRP492079</td>\n",
       "      <td>SRX23767835</td>\n",
       "      <td>Nanopore DNA sequencing of Equus ferus przewal...</td>\n",
       "      <td>9798</td>\n",
       "      <td>Equus przewalskii</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS20593331</td>\n",
       "      <td>Varuska</td>\n",
       "      <td>PromethION</td>\n",
       "      <td>13536413</td>\n",
       "      <td>93287592276</td>\n",
       "      <td>SRR28123101</td>\n",
       "      <td>13536413</td>\n",
       "      <td>111175141362</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>626 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_accession experiment_accession  \\\n",
       "0         SRP500767          SRX24214614   \n",
       "1         SRP499270          SRX24128005   \n",
       "2         SRP496639          SRX24019922   \n",
       "3         SRP496639          SRX24019921   \n",
       "4         SRP496639          SRX24019920   \n",
       "..              ...                  ...   \n",
       "621       SRP417920          SRX19052651   \n",
       "622       SRP413991          SRX18782374   \n",
       "623       SRP413991          SRX18782373   \n",
       "624       SRP413936          SRX18776194   \n",
       "625       SRP492079          SRX23767835   \n",
       "\n",
       "                                      experiment_title sample_taxon_id  \\\n",
       "0                                     ONT seq of Pf_T2          463681   \n",
       "1    Basecalled ONT Sequences of Spirobacillus cien...          495820   \n",
       "2     ONT sequencing of Staphylococcus aureus TUM20914            1280   \n",
       "3     ONT sequencing of Staphylococcus aureus TUM20913            1280   \n",
       "4     ONT sequencing of Staphylococcus aureus TUM20888            1280   \n",
       "..                                                 ...             ...   \n",
       "621  Whole Genome Sequencing of three Canadian stra...             562   \n",
       "622  Oxford Nanopore Q20 400bps sequencing of adult...            7227   \n",
       "623  Oxford Nanopore Q20 260bps sequencing of adult...            7227   \n",
       "624                        Madurella fahalii CBS129176         1157608   \n",
       "625  Nanopore DNA sequencing of Equus ferus przewal...            9798   \n",
       "\n",
       "        sample_scientific_name experiment_library_strategy  \\\n",
       "0     Penicillium fuscoglaucum                         WGS   \n",
       "1    Spirobacillus cienkowskii                         WGS   \n",
       "2        Staphylococcus aureus                         WGS   \n",
       "3        Staphylococcus aureus                         WGS   \n",
       "4        Staphylococcus aureus                         WGS   \n",
       "..                         ...                         ...   \n",
       "621           Escherichia coli                         WGS   \n",
       "622    Drosophila melanogaster                         WGS   \n",
       "623    Drosophila melanogaster                         WGS   \n",
       "624          Madurella fahalii                         WGS   \n",
       "625          Equus przewalskii                         WGS   \n",
       "\n",
       "    experiment_library_source experiment_library_selection sample_accession  \\\n",
       "0                     GENOMIC           size fractionation      SRS20984821   \n",
       "1                     GENOMIC                       RANDOM      SRS20913379   \n",
       "2                     GENOMIC                       RANDOM      SRS20798084   \n",
       "3                     GENOMIC                       RANDOM      SRS20798083   \n",
       "4                     GENOMIC                       RANDOM      SRS20798082   \n",
       "..                        ...                          ...              ...   \n",
       "621                   GENOMIC                       RANDOM      SRS16469023   \n",
       "622                   GENOMIC                       RANDOM      SRS16216412   \n",
       "623                   GENOMIC                       RANDOM      SRS16216412   \n",
       "624                   GENOMIC                       RANDOM      SRS16210388   \n",
       "625                   GENOMIC                       RANDOM      SRS20593331   \n",
       "\n",
       "                         sample_alias experiment_instrument_model  \\\n",
       "0                               Pf_T2                      MinION   \n",
       "1    Spirobacillus_cienkowskii_FI-SP1                      MinION   \n",
       "2      Staphylococcus aureus TUM20914                      MinION   \n",
       "3      Staphylococcus aureus TUM20913                      MinION   \n",
       "4      Staphylococcus aureus TUM20888                      MinION   \n",
       "..                                ...                         ...   \n",
       "621                              HH45                      MinION   \n",
       "622                dmel-bdgp-iso1-Q20                      MinION   \n",
       "623                dmel-bdgp-iso1-Q20                      MinION   \n",
       "624                         CBS129176                      MinION   \n",
       "625                           Varuska                  PromethION   \n",
       "\n",
       "    pool_member_spots   run_1_size run_1_accession run_1_total_spots  \\\n",
       "0             1723079  13484052567     SRR28616354           1723079   \n",
       "1             5283013   3913249236     SRR28527199           5283013   \n",
       "2               69084    340431788     SRR28415513             69084   \n",
       "3              184459    681816343     SRR28415514            184459   \n",
       "4              209063   1235881070     SRR28415515            209063   \n",
       "..                ...          ...             ...               ...   \n",
       "621            251505    376551024     SRR23100674            251505   \n",
       "622           1937835  18615677517     SRR22822929           1937835   \n",
       "623           1481121  16566954959     SRR22822930           1481121   \n",
       "624            696330   2085105473     SRR22816638            696330   \n",
       "625          13536413  93287592276     SRR28123101          13536413   \n",
       "\n",
       "    run_1_total_bases  \n",
       "0         15302485972  \n",
       "1          4450905037  \n",
       "2           368340750  \n",
       "3           735944460  \n",
       "4          1349673232  \n",
       "..                ...  \n",
       "621         437387806  \n",
       "622       21140714134  \n",
       "623       19204298459  \n",
       "624        2327283485  \n",
       "625      111175141362  \n",
       "\n",
       "[626 rows x 16 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>sample_taxon_id</th>\n",
       "      <th>sample_scientific_name</th>\n",
       "      <th>experiment_library_strategy</th>\n",
       "      <th>experiment_library_source</th>\n",
       "      <th>experiment_library_selection</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>experiment_instrument_model</th>\n",
       "      <th>pool_member_spots</th>\n",
       "      <th>run_1_size</th>\n",
       "      <th>run_1_accession</th>\n",
       "      <th>run_1_total_spots</th>\n",
       "      <th>run_1_total_bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>SRP302935</td>\n",
       "      <td>SRX23621084</td>\n",
       "      <td>STEC O157</td>\n",
       "      <td>1005487</td>\n",
       "      <td>Escherichia coli PA2</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS355621</td>\n",
       "      <td>PO157_PA2</td>\n",
       "      <td>GridION</td>\n",
       "      <td>89920</td>\n",
       "      <td>604308431</td>\n",
       "      <td>SRR27967420</td>\n",
       "      <td>89920</td>\n",
       "      <td>712881969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>SRP302935</td>\n",
       "      <td>SRX23621083</td>\n",
       "      <td>STEC O157</td>\n",
       "      <td>1005492</td>\n",
       "      <td>Escherichia coli PA8</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS355432</td>\n",
       "      <td>PO157_PA8</td>\n",
       "      <td>GridION</td>\n",
       "      <td>110329</td>\n",
       "      <td>666476517</td>\n",
       "      <td>SRR27967421</td>\n",
       "      <td>110329</td>\n",
       "      <td>783471849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>611</th>\n",
       "      <td>SRP418927</td>\n",
       "      <td>SRX19161994</td>\n",
       "      <td>Nanopore R10.4.1 seq of adult muscle</td>\n",
       "      <td>10060</td>\n",
       "      <td>Ondatra zibethicus</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS16572737</td>\n",
       "      <td>Muskrat</td>\n",
       "      <td>MinION</td>\n",
       "      <td>56048</td>\n",
       "      <td>48740656</td>\n",
       "      <td>SRR23214093</td>\n",
       "      <td>56048</td>\n",
       "      <td>38552743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553</th>\n",
       "      <td>SRP418927</td>\n",
       "      <td>SRX19162031</td>\n",
       "      <td>Nanopore R10.4.1 seq of synthetic control</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS16572774</td>\n",
       "      <td>17_Mouse_low</td>\n",
       "      <td>MinION</td>\n",
       "      <td>67159</td>\n",
       "      <td>88101059</td>\n",
       "      <td>SRR23214056</td>\n",
       "      <td>67159</td>\n",
       "      <td>117450679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>SRP418927</td>\n",
       "      <td>SRX19161992</td>\n",
       "      <td>Nanopore R10.4.1 seq of adult muscle</td>\n",
       "      <td>10116</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS16572735</td>\n",
       "      <td>Rat</td>\n",
       "      <td>MinION</td>\n",
       "      <td>22681</td>\n",
       "      <td>11949294</td>\n",
       "      <td>SRR23214095</td>\n",
       "      <td>22681</td>\n",
       "      <td>10519941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>SRP418927</td>\n",
       "      <td>SRX19162037</td>\n",
       "      <td>Nanopore R10.4.1 seq of adult muscle</td>\n",
       "      <td>9925</td>\n",
       "      <td>Capra hircus</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS16572780</td>\n",
       "      <td>Goat</td>\n",
       "      <td>MinION</td>\n",
       "      <td>97784</td>\n",
       "      <td>65786578</td>\n",
       "      <td>SRR23214050</td>\n",
       "      <td>97784</td>\n",
       "      <td>82269704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>SRP418927</td>\n",
       "      <td>SRX19162026</td>\n",
       "      <td>Nanopore R10.4.1 seq of adult muscle</td>\n",
       "      <td>9940</td>\n",
       "      <td>Ovis aries</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS16572769</td>\n",
       "      <td>Sheep</td>\n",
       "      <td>MinION</td>\n",
       "      <td>44511</td>\n",
       "      <td>53369602</td>\n",
       "      <td>SRR23214061</td>\n",
       "      <td>44511</td>\n",
       "      <td>66487309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>SRP418927</td>\n",
       "      <td>SRX19161990</td>\n",
       "      <td>Nanopore R10.4.1 seq of adult muscle</td>\n",
       "      <td>9986</td>\n",
       "      <td>Oryctolagus cuniculus</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS16572733</td>\n",
       "      <td>Rabbit</td>\n",
       "      <td>MinION</td>\n",
       "      <td>97572</td>\n",
       "      <td>110935136</td>\n",
       "      <td>SRR23214097</td>\n",
       "      <td>97572</td>\n",
       "      <td>147850885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>SRP484228</td>\n",
       "      <td>SRX23262812</td>\n",
       "      <td>gDNA D. dudresnayi gametophyte ONT</td>\n",
       "      <td>99931</td>\n",
       "      <td>Desmarestia dudresnayi</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS20164780</td>\n",
       "      <td>D_dudresnayi_sxs43</td>\n",
       "      <td>MinION</td>\n",
       "      <td>1908772</td>\n",
       "      <td>2556008277</td>\n",
       "      <td>SRR27594101</td>\n",
       "      <td>1908772</td>\n",
       "      <td>9785795104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>SRP418927</td>\n",
       "      <td>SRX19161995</td>\n",
       "      <td>Nanopore R10.4.1 seq of adult muscle</td>\n",
       "      <td>9995</td>\n",
       "      <td>Marmota monax</td>\n",
       "      <td>WGS</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>SRS16572738</td>\n",
       "      <td>Woodchuck</td>\n",
       "      <td>MinION</td>\n",
       "      <td>88742</td>\n",
       "      <td>605240888</td>\n",
       "      <td>SRR23214092</td>\n",
       "      <td>88742</td>\n",
       "      <td>83519273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>286 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_accession experiment_accession  \\\n",
       "47        SRP302935          SRX23621084   \n",
       "48        SRP302935          SRX23621083   \n",
       "611       SRP418927          SRX19161994   \n",
       "553       SRP418927          SRX19162031   \n",
       "568       SRP418927          SRX19161992   \n",
       "..              ...                  ...   \n",
       "585       SRP418927          SRX19162037   \n",
       "555       SRP418927          SRX19162026   \n",
       "613       SRP418927          SRX19161990   \n",
       "155       SRP484228          SRX23262812   \n",
       "567       SRP418927          SRX19161995   \n",
       "\n",
       "                              experiment_title sample_taxon_id  \\\n",
       "47                                   STEC O157         1005487   \n",
       "48                                   STEC O157         1005492   \n",
       "611       Nanopore R10.4.1 seq of adult muscle           10060   \n",
       "553  Nanopore R10.4.1 seq of synthetic control           10090   \n",
       "568       Nanopore R10.4.1 seq of adult muscle           10116   \n",
       "..                                         ...             ...   \n",
       "585       Nanopore R10.4.1 seq of adult muscle            9925   \n",
       "555       Nanopore R10.4.1 seq of adult muscle            9940   \n",
       "613       Nanopore R10.4.1 seq of adult muscle            9986   \n",
       "155         gDNA D. dudresnayi gametophyte ONT           99931   \n",
       "567       Nanopore R10.4.1 seq of adult muscle            9995   \n",
       "\n",
       "     sample_scientific_name experiment_library_strategy  \\\n",
       "47     Escherichia coli PA2                         WGS   \n",
       "48     Escherichia coli PA8                         WGS   \n",
       "611      Ondatra zibethicus                         WGS   \n",
       "553            Mus musculus                         WGS   \n",
       "568       Rattus norvegicus                         WGS   \n",
       "..                      ...                         ...   \n",
       "585            Capra hircus                         WGS   \n",
       "555              Ovis aries                         WGS   \n",
       "613   Oryctolagus cuniculus                         WGS   \n",
       "155  Desmarestia dudresnayi                         WGS   \n",
       "567           Marmota monax                         WGS   \n",
       "\n",
       "    experiment_library_source experiment_library_selection sample_accession  \\\n",
       "47                    GENOMIC                       RANDOM        SRS355621   \n",
       "48                    GENOMIC                       RANDOM        SRS355432   \n",
       "611                   GENOMIC                       RANDOM      SRS16572737   \n",
       "553                   GENOMIC                       RANDOM      SRS16572774   \n",
       "568                   GENOMIC                       RANDOM      SRS16572735   \n",
       "..                        ...                          ...              ...   \n",
       "585                   GENOMIC                       RANDOM      SRS16572780   \n",
       "555                   GENOMIC                       RANDOM      SRS16572769   \n",
       "613                   GENOMIC                       RANDOM      SRS16572733   \n",
       "155                   GENOMIC                       RANDOM      SRS20164780   \n",
       "567                   GENOMIC                       RANDOM      SRS16572738   \n",
       "\n",
       "           sample_alias experiment_instrument_model pool_member_spots  \\\n",
       "47            PO157_PA2                     GridION             89920   \n",
       "48            PO157_PA8                     GridION            110329   \n",
       "611             Muskrat                      MinION             56048   \n",
       "553        17_Mouse_low                      MinION             67159   \n",
       "568                 Rat                      MinION             22681   \n",
       "..                  ...                         ...               ...   \n",
       "585                Goat                      MinION             97784   \n",
       "555               Sheep                      MinION             44511   \n",
       "613              Rabbit                      MinION             97572   \n",
       "155  D_dudresnayi_sxs43                      MinION           1908772   \n",
       "567           Woodchuck                      MinION             88742   \n",
       "\n",
       "     run_1_size run_1_accession run_1_total_spots run_1_total_bases  \n",
       "47    604308431     SRR27967420             89920         712881969  \n",
       "48    666476517     SRR27967421            110329         783471849  \n",
       "611    48740656     SRR23214093             56048          38552743  \n",
       "553    88101059     SRR23214056             67159         117450679  \n",
       "568    11949294     SRR23214095             22681          10519941  \n",
       "..          ...             ...               ...               ...  \n",
       "585    65786578     SRR23214050             97784          82269704  \n",
       "555    53369602     SRR23214061             44511          66487309  \n",
       "613   110935136     SRR23214097             97572         147850885  \n",
       "155  2556008277     SRR27594101           1908772        9785795104  \n",
       "567   605240888     SRR23214092             88742          83519273  \n",
       "\n",
       "[286 rows x 16 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df = res_df.groupby(\"sample_taxon_id\").sample(1)\n",
    "sample_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 22%|██▏       | 63/286 [16:53<1:19:33, 21.41s/it]"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "run_sample_info_df = None\n",
    "\n",
    "for i in tqdm(range(len(sample_df))):\n",
    "    study_df = db.sra_metadata(sample_df.iloc[i][\"study_accession\"], detailed=True)\n",
    "    study_df_sample = study_df.sample(1)\n",
    "\n",
    "    if run_sample_info_df is None:\n",
    "        run_sample_info_df = study_df_sample\n",
    "    else:\n",
    "        run_sample_info_df = pd.concat([run_sample_info_df, study_df_sample])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_sample_info_df = run_sample_info_df[run_sample_info_df[\"instrument_model_desc\"] == \"OXFORD_NANOPORE\"]\n",
    "run_sample_info_df.to_csv(\"nanopore_10.4_data_samples.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-04-01T08:12:19 prefetch.2.11.3 int: libs/vfs/path.c:1820:VPathParseInt: character unexpected while parsing path within virtual file system module - VFSManagerMakePath: 0    SRR25865495 0    SRR28421270 0    SRR23879687 0    SRR25523715 0    SRR24446466 0    SRR21006450 0    SRR24446466 Name: run_accession, dtype: object\n",
      "2024-04-01T08:12:19 prefetch.2.11.3 int: libs/vfs/path.c:1820:VPathParseInt: character unexpected while parsing path within virtual file system module - VFSManagerMakePath: 1    SRR26109698 1    SRR25581455 1    SRR27467590 1    SRR23693644 Name: run_accession, dtype: object\n",
      "2024-04-01T08:12:19 prefetch.2.11.3 int: libs/vfs/path.c:1820:VPathParseInt: character unexpected while parsing path within virtual file system module - VFSManagerMakePath: 2    SRR27467479 2    SRR27467450 2    SRR26049867 2    SRR24446468 Name: run_accession, dtype: object\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-04-01T08:12:21 prefetch.2.11.3: Current preference is set to retrieve SRA Normalized Format files with full base quality scores.\n",
      "2024-04-01T08:12:23 prefetch.2.11.3: 1) Downloading 'SRR24446469'...\n",
      "2024-04-01T08:12:23 prefetch.2.11.3: SRA Normalized Format file is being retrieved, if this is different from your preference, it may be due to current file availability.\n",
      "2024-04-01T08:12:23 prefetch.2.11.3:  Downloading via HTTPS...\n",
      "2024-04-01T09:00:28 prefetch.2.11.3:  HTTPS download succeed\n",
      "2024-04-01T09:00:35 prefetch.2.11.3:  'SRR24446469' is valid\n",
      "2024-04-01T09:00:35 prefetch.2.11.3: 1) 'SRR24446469' was downloaded successfully\n",
      "2024-04-01T09:00:35 prefetch.2.11.3: 'SRR24446469' has 0 unresolved dependencies\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-04-01T09:00:38 prefetch.2.11.3: Current preference is set to retrieve SRA Normalized Format files with full base quality scores.\n",
      "2024-04-01T09:00:40 prefetch.2.11.3: 1) Downloading 'SRR24446470'...\n",
      "2024-04-01T09:00:40 prefetch.2.11.3: SRA Normalized Format file is being retrieved, if this is different from your preference, it may be due to current file availability.\n",
      "2024-04-01T09:00:40 prefetch.2.11.3:  Downloading via HTTPS...\n",
      "2024-04-01T10:03:27 prefetch.2.11.3:  HTTPS download succeed\n",
      "2024-04-01T10:03:36 prefetch.2.11.3:  'SRR24446470' is valid\n",
      "2024-04-01T10:03:36 prefetch.2.11.3: 1) 'SRR24446470' was downloaded successfully\n",
      "2024-04-01T10:03:36 prefetch.2.11.3: 'SRR24446470' has 0 unresolved dependencies\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-04-01T10:03:38 prefetch.2.11.3: Current preference is set to retrieve SRA Normalized Format files with full base quality scores.\n",
      "2024-04-01T10:03:40 prefetch.2.11.3 warn: Maximum file size download limit is 20GB \n",
      "2024-04-01T10:03:40 prefetch.2.11.3: 1) 'SRR21397272' (22GB) is larger than maximum allowed: skipped \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Download of some files was skipped because they are too large\n",
      "You can change size download limit by setting\n",
      "--min-size and --max-size command line arguments\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "6",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/indexes/base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3802\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:153\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:177\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:220\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine._get_loc_duplicates\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:228\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine._maybe_get_bool_indexer\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:120\u001b[0m, in \u001b[0;36mpandas._libs.index._unpack_bool_indexer\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 6",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[53], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m os\u001b[38;5;241m.\u001b[39mchdir(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m./data\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(run_sample_info_df)):\n\u001b[0;32m----> 5\u001b[0m     subprocess\u001b[38;5;241m.\u001b[39mrun([\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprefetch\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mstr\u001b[39m(\u001b[43mrun_sample_info_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mi\u001b[49m\u001b[43m]\u001b[49m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrun_accession\u001b[39m\u001b[38;5;124m\"\u001b[39m])])\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/indexing.py:1192\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1190\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m com\u001b[38;5;241m.\u001b[39mapply_if_callable(key, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj)\n\u001b[1;32m   1191\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_deprecated_callable_usage(key, maybe_callable)\n\u001b[0;32m-> 1192\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaybe_callable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/indexing.py:1432\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1430\u001b[0m \u001b[38;5;66;03m# fall thru to straight lookup\u001b[39;00m\n\u001b[1;32m   1431\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_key(key, axis)\n\u001b[0;32m-> 1432\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_label\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/indexing.py:1382\u001b[0m, in \u001b[0;36m_LocIndexer._get_label\u001b[0;34m(self, label, axis)\u001b[0m\n\u001b[1;32m   1380\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_get_label\u001b[39m(\u001b[38;5;28mself\u001b[39m, label, axis: AxisInt):\n\u001b[1;32m   1381\u001b[0m     \u001b[38;5;66;03m# GH#5567 this will fail if the label is not present in the axis.\u001b[39;00m\n\u001b[0;32m-> 1382\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mxs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/generic.py:4295\u001b[0m, in \u001b[0;36mNDFrame.xs\u001b[0;34m(self, key, axis, level, drop_level)\u001b[0m\n\u001b[1;32m   4293\u001b[0m             new_index \u001b[38;5;241m=\u001b[39m index[loc]\n\u001b[1;32m   4294\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 4295\u001b[0m     loc \u001b[38;5;241m=\u001b[39m \u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4297\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(loc, np\u001b[38;5;241m.\u001b[39mndarray):\n\u001b[1;32m   4298\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m loc\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m np\u001b[38;5;241m.\u001b[39mbool_:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/indexes/base.py:3809\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3805\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3806\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3807\u001b[0m     ):\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3809\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3810\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3811\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3812\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 6"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "\n",
    "num_samples = 1000\n",
    "\n",
    "os.chdir(\"./data\")\n",
    "for i in range(len(run_sample_info_df)):\n",
    "    subprocess.run([\"fastq-dump\", str(run_sample_info_df.loc[i][\"run_accession\"]), \"-X\", str(num_samples)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbi_species_taxid</th>\n",
       "      <th>ncbi_taxid</th>\n",
       "      <th>ncbi_organism_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1313</td>\n",
       "      <td>1313</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1313</td>\n",
       "      <td>1313</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>216816</td>\n",
       "      <td>1457183</td>\n",
       "      <td>Bifidobacterium longum subsp. infantis EK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>661</td>\n",
       "      <td>661</td>\n",
       "      <td>Photobacterium angustum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1359</td>\n",
       "      <td>1295826</td>\n",
       "      <td>Lactococcus lactis subsp. cremoris KW2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ncbi_species_taxid  ncbi_taxid                          ncbi_organism_name\n",
       "0                1313        1313                    Streptococcus pneumoniae\n",
       "1                1313        1313                    Streptococcus pneumoniae\n",
       "2              216816     1457183  Bifidobacterium longum subsp. infantis EK3\n",
       "3                 661         661                     Photobacterium angustum\n",
       "4                1359     1295826      Lactococcus lactis subsp. cremoris KW2"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_metadata.head()[[\"ncbi_species_taxid\", \"ncbi_taxid\", \"ncbi_organism_name\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "57939"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_set = set(gtdb_metadata[\"ncbi_species_taxid\"])\n",
    "len(species_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtdb_metadata.loc[0][\"ncbi_organism_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[22], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43mdb\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msearch_sra\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgtdb_metadata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mncbi_organism_name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pysradb/sradb.py:1172\u001b[0m, in \u001b[0;36mSRAdb.search_sra\u001b[0;34m(self, search_str, out_type, assay, sample_attribute, detailed, expand_sample_attributes, output_read_lengths)\u001b[0m\n\u001b[1;32m   1142\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msearch_sra\u001b[39m(\n\u001b[1;32m   1143\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1144\u001b[0m     search_str,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1155\u001b[0m     output_read_lengths\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m   1156\u001b[0m ):\n\u001b[1;32m   1157\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Search SRA for any search term.\u001b[39;00m\n\u001b[1;32m   1158\u001b[0m \n\u001b[1;32m   1159\u001b[0m \u001b[38;5;124;03m    Parameters\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1170\u001b[0m \u001b[38;5;124;03m              Dataframe with relevant query results\u001b[39;00m\n\u001b[1;32m   1171\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1172\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msra_metadata\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1173\u001b[0m \u001b[43m        \u001b[49m\u001b[43msearch_str\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1174\u001b[0m \u001b[43m        \u001b[49m\u001b[43mout_type\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mout_type\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1175\u001b[0m \u001b[43m        \u001b[49m\u001b[43massay\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43massay\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1176\u001b[0m \u001b[43m        \u001b[49m\u001b[43msample_attribute\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msample_attribute\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1177\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdetailed\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdetailed\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1178\u001b[0m \u001b[43m        \u001b[49m\u001b[43mexpand_sample_attributes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexpand_sample_attributes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1179\u001b[0m \u001b[43m        \u001b[49m\u001b[43moutput_read_lengths\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_read_lengths\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1180\u001b[0m \u001b[43m        \u001b[49m\u001b[43macc_is_searchstr\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m   1181\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pysradb/sraweb.py:415\u001b[0m, in \u001b[0;36mSRAweb.sra_metadata\u001b[0;34m(self, srp, sample_attribute, detailed, expand_sample_attributes, output_read_lengths, **kwargs)\u001b[0m\n\u001b[1;32m    406\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msra_metadata\u001b[39m(\n\u001b[1;32m    407\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    408\u001b[0m     srp,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    413\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs\n\u001b[1;32m    414\u001b[0m ):\n\u001b[0;32m--> 415\u001b[0m     esummary_result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_esummary_response\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msra\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msrp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    416\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    417\u001b[0m         uids \u001b[38;5;241m=\u001b[39m esummary_result[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muids\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pysradb/sraweb.py:298\u001b[0m, in \u001b[0;36mSRAweb.get_esummary_response\u001b[0;34m(self, db, term, usehistory)\u001b[0m\n\u001b[1;32m    296\u001b[0m payload \u001b[38;5;241m=\u001b[39m OrderedDict(payload)\n\u001b[1;32m    297\u001b[0m payload[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mretstart\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m retstart\n\u001b[0;32m--> 298\u001b[0m request \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    299\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbase_url\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mesummary\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mOrderedDict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpayload\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    300\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    301\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    302\u001b[0m     response \u001b[38;5;241m=\u001b[39m request\u001b[38;5;241m.\u001b[39mjson()\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/requests/sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    706\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/requests/adapters.py:486\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    483\u001b[0m     timeout \u001b[38;5;241m=\u001b[39m TimeoutSauce(connect\u001b[38;5;241m=\u001b[39mtimeout, read\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    485\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 486\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    487\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    488\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    489\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    490\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    491\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    492\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    495\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    496\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    497\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    498\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    500\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ProtocolError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    501\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(err, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:715\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    712\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_prepare_proxy(conn)\n\u001b[1;32m    714\u001b[0m \u001b[38;5;66;03m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 715\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    716\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    717\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    718\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    719\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    720\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    721\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    722\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    723\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    725\u001b[0m \u001b[38;5;66;03m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    726\u001b[0m \u001b[38;5;66;03m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    727\u001b[0m \u001b[38;5;66;03m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    728\u001b[0m \u001b[38;5;66;03m# mess.\u001b[39;00m\n\u001b[1;32m    729\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:404\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    402\u001b[0m \u001b[38;5;66;03m# Trigger any extra validation we need to do.\u001b[39;00m\n\u001b[1;32m    403\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 404\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    405\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    406\u001b[0m     \u001b[38;5;66;03m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[39;00m\n\u001b[1;32m    407\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mconn\u001b[38;5;241m.\u001b[39mtimeout)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/connectionpool.py:1058\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   1056\u001b[0m \u001b[38;5;66;03m# Force connect early to allow us to validate the connection.\u001b[39;00m\n\u001b[1;32m   1057\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(conn, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msock\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):  \u001b[38;5;66;03m# AppEngine might not have  `.sock`\u001b[39;00m\n\u001b[0;32m-> 1058\u001b[0m     \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1060\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_verified:\n\u001b[1;32m   1061\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m   1062\u001b[0m         (\n\u001b[1;32m   1063\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnverified HTTPS request is being made to host \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1068\u001b[0m         InsecureRequestWarning,\n\u001b[1;32m   1069\u001b[0m     )\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/connection.py:363\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    361\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mconnect\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    362\u001b[0m     \u001b[38;5;66;03m# Add certificate verification\u001b[39;00m\n\u001b[0;32m--> 363\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_new_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    364\u001b[0m     hostname \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost\n\u001b[1;32m    365\u001b[0m     tls_in_tls \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/connection.py:174\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    171\u001b[0m     extra_kw[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msocket_options\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msocket_options\n\u001b[1;32m    173\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 174\u001b[0m     conn \u001b[38;5;241m=\u001b[39m \u001b[43mconnection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_connection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    175\u001b[0m \u001b[43m        \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_dns_host\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mport\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mextra_kw\u001b[49m\n\u001b[1;32m    176\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    178\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m SocketTimeout:\n\u001b[1;32m    179\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m ConnectTimeoutError(\n\u001b[1;32m    180\u001b[0m         \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    181\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mConnection to \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m timed out. (connect timeout=\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    182\u001b[0m         \u001b[38;5;241m%\u001b[39m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtimeout),\n\u001b[1;32m    183\u001b[0m     )\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/urllib3/util/connection.py:85\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address, socket_options)\u001b[0m\n\u001b[1;32m     83\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m source_address:\n\u001b[1;32m     84\u001b[0m         sock\u001b[38;5;241m.\u001b[39mbind(source_address)\n\u001b[0;32m---> 85\u001b[0m     \u001b[43msock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43msa\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     86\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m sock\n\u001b[1;32m     88\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m socket\u001b[38;5;241m.\u001b[39merror \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "res = db.search_sra(gtdb_metadata.loc[0][\"ncbi_organism_name\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>study_title</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>experiment_desc</th>\n",
       "      <th>organism_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>...</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>type-material</th>\n",
       "      <th>biosamplemodel</th>\n",
       "      <th>sub_species</th>\n",
       "      <th>ena_fastq_http</th>\n",
       "      <th>ena_fastq_http_1</th>\n",
       "      <th>ena_fastq_http_2</th>\n",
       "      <th>ena_fastq_ftp</th>\n",
       "      <th>ena_fastq_ftp_1</th>\n",
       "      <th>ena_fastq_ftp_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR24651172</td>\n",
       "      <td>SRP438478</td>\n",
       "      <td>Genome sequencing and hybrid assembly of micro...</td>\n",
       "      <td>SRX20431359</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>537012</td>\n",
       "      <td>Bacteroides cellulosilyticus DSM 14838</td>\n",
       "      <td>Nanopore-sequencing-of-Bacteroides-cellulosily...</td>\n",
       "      <td>WGS</td>\n",
       "      <td>...</td>\n",
       "      <td>not collected</td>\n",
       "      <td>type strain of Bacteroides cellulosilyticus</td>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR24651174</td>\n",
       "      <td>SRP438478</td>\n",
       "      <td>Genome sequencing and hybrid assembly of micro...</td>\n",
       "      <td>SRX20431357</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>84112</td>\n",
       "      <td>Eggerthella lenta</td>\n",
       "      <td>Nanopore-sequencing-of-Eggerthella-lenta-DSM-2243</td>\n",
       "      <td>WGS</td>\n",
       "      <td>...</td>\n",
       "      <td>not collected</td>\n",
       "      <td>type strain of Eggerthella lenta</td>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR24651176</td>\n",
       "      <td>SRP438478</td>\n",
       "      <td>Genome sequencing and hybrid assembly of micro...</td>\n",
       "      <td>SRX20431355</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>338188</td>\n",
       "      <td>Bacteroides finegoldii</td>\n",
       "      <td>Nanopore-sequencing-of-Bacteroides-finegoldii-...</td>\n",
       "      <td>WGS</td>\n",
       "      <td>...</td>\n",
       "      <td>not collected</td>\n",
       "      <td>type strain of Bacteroides finegoldii</td>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR24651177</td>\n",
       "      <td>SRP438478</td>\n",
       "      <td>Genome sequencing and hybrid assembly of micro...</td>\n",
       "      <td>SRX20431354</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>435591</td>\n",
       "      <td>Parabacteroides distasonis ATCC 8503</td>\n",
       "      <td>Nanopore-sequencing-of-Parabacteroides-distaso...</td>\n",
       "      <td>WGS</td>\n",
       "      <td>...</td>\n",
       "      <td>not collected</td>\n",
       "      <td>type strain of Parabacteroides distasonis</td>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR24651179</td>\n",
       "      <td>SRP438478</td>\n",
       "      <td>Genome sequencing and hybrid assembly of micro...</td>\n",
       "      <td>SRX20431352</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>PCR free Oxford Nanopore sequencing of genomic...</td>\n",
       "      <td>84030</td>\n",
       "      <td>Lacrimispora saccharolytica</td>\n",
       "      <td>Nanopore-sequencing-of-Lacrimispora-saccharoly...</td>\n",
       "      <td>WGS</td>\n",
       "      <td>...</td>\n",
       "      <td>not collected</td>\n",
       "      <td>type strain of Clostridium saccharolyticum</td>\n",
       "      <td>MIGS/MIMS/MIMARKS.human-gut</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession study_accession  \\\n",
       "1   SRR24651172       SRP438478   \n",
       "3   SRR24651174       SRP438478   \n",
       "5   SRR24651176       SRP438478   \n",
       "6   SRR24651177       SRP438478   \n",
       "8   SRR24651179       SRP438478   \n",
       "\n",
       "                                         study_title experiment_accession  \\\n",
       "1  Genome sequencing and hybrid assembly of micro...          SRX20431359   \n",
       "3  Genome sequencing and hybrid assembly of micro...          SRX20431357   \n",
       "5  Genome sequencing and hybrid assembly of micro...          SRX20431355   \n",
       "6  Genome sequencing and hybrid assembly of micro...          SRX20431354   \n",
       "8  Genome sequencing and hybrid assembly of micro...          SRX20431352   \n",
       "\n",
       "                                    experiment_title  \\\n",
       "1  PCR free Oxford Nanopore sequencing of genomic...   \n",
       "3  PCR free Oxford Nanopore sequencing of genomic...   \n",
       "5  PCR free Oxford Nanopore sequencing of genomic...   \n",
       "6  PCR free Oxford Nanopore sequencing of genomic...   \n",
       "8  PCR free Oxford Nanopore sequencing of genomic...   \n",
       "\n",
       "                                     experiment_desc organism_taxid  \\\n",
       "1  PCR free Oxford Nanopore sequencing of genomic...         537012   \n",
       "3  PCR free Oxford Nanopore sequencing of genomic...          84112   \n",
       "5  PCR free Oxford Nanopore sequencing of genomic...         338188   \n",
       "6  PCR free Oxford Nanopore sequencing of genomic...         435591   \n",
       "8  PCR free Oxford Nanopore sequencing of genomic...          84030   \n",
       "\n",
       "                            organism_name  \\\n",
       "1  Bacteroides cellulosilyticus DSM 14838   \n",
       "3                       Eggerthella lenta   \n",
       "5                  Bacteroides finegoldii   \n",
       "6    Parabacteroides distasonis ATCC 8503   \n",
       "8             Lacrimispora saccharolytica   \n",
       "\n",
       "                                        library_name library_strategy  ...  \\\n",
       "1  Nanopore-sequencing-of-Bacteroides-cellulosily...              WGS  ...   \n",
       "3  Nanopore-sequencing-of-Eggerthella-lenta-DSM-2243              WGS  ...   \n",
       "5  Nanopore-sequencing-of-Bacteroides-finegoldii-...              WGS  ...   \n",
       "6  Nanopore-sequencing-of-Parabacteroides-distaso...              WGS  ...   \n",
       "8  Nanopore-sequencing-of-Lacrimispora-saccharoly...              WGS  ...   \n",
       "\n",
       "  ref_biomaterial                                type-material  \\\n",
       "1   not collected  type strain of Bacteroides cellulosilyticus   \n",
       "3   not collected             type strain of Eggerthella lenta   \n",
       "5   not collected        type strain of Bacteroides finegoldii   \n",
       "6   not collected    type strain of Parabacteroides distasonis   \n",
       "8   not collected   type strain of Clostridium saccharolyticum   \n",
       "\n",
       "                biosamplemodel sub_species ena_fastq_http ena_fastq_http_1  \\\n",
       "1  MIGS/MIMS/MIMARKS.human-gut        <NA>           <NA>             <NA>   \n",
       "3  MIGS/MIMS/MIMARKS.human-gut        <NA>           <NA>             <NA>   \n",
       "5  MIGS/MIMS/MIMARKS.human-gut        <NA>           <NA>             <NA>   \n",
       "6  MIGS/MIMS/MIMARKS.human-gut        <NA>           <NA>             <NA>   \n",
       "8  MIGS/MIMS/MIMARKS.human-gut        <NA>           <NA>             <NA>   \n",
       "\n",
       "  ena_fastq_http_2 ena_fastq_ftp ena_fastq_ftp_1 ena_fastq_ftp_2  \n",
       "1             <NA>          <NA>            <NA>            <NA>  \n",
       "3             <NA>          <NA>            <NA>            <NA>  \n",
       "5             <NA>          <NA>            <NA>            <NA>  \n",
       "6             <NA>          <NA>            <NA>            <NA>  \n",
       "8             <NA>          <NA>            <NA>            <NA>  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_df = db.sra_metadata(\"SRP438478\", detailed=True)\n",
    "\n",
    "study_df = study_df[study_df[\"instrument_model_desc\"] == \"OXFORD_NANOPORE\"]\n",
    "study_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR24651172\n",
      "SRP438478\n",
      "Genome sequencing and hybrid assembly of microbial isolates derived from the human gut.\n",
      "SRX20431359\n",
      "PCR free Oxford Nanopore sequencing of genomic DNA from Bacteroides-cellulosilyticus-DSM-14838\n",
      "PCR free Oxford Nanopore sequencing of genomic DNA from Bacteroides-cellulosilyticus-DSM-14838\n",
      "537012\n",
      "Bacteroides cellulosilyticus DSM 14838\n",
      "Nanopore-sequencing-of-Bacteroides-cellulosilyticus-DSM-14838\n",
      "WGS\n",
      "GENOMIC\n",
      "RANDOM\n",
      "SINGLE\n",
      "SRS17736875\n",
      "<NA>\n",
      "SAMN34894971\n",
      "PRJNA885826\n",
      "MinION\n",
      "MinION\n",
      "OXFORD_NANOPORE\n",
      "178241\n",
      "2652210907\n",
      "178241\n",
      "2938492130\n",
      "Bacteroides-cellulosilyticus-DSM-14838__barcode15.fastq.gz\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "https://sra-pub-run-odp.s3.amazonaws.com/sra/SRR24651172/SRR24651172\n",
      "worldwide\n",
      "anonymous\n",
      "<NA>\n",
      "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos3/sra-pub-run-36/SRR024/24651/SRR24651172/SRR24651172.1\n",
      "worldwide\n",
      "anonymous\n",
      "gs://sra-pub-run-13/SRR24651172/SRR24651172.1\n",
      "gs.US\n",
      "gcp identity\n",
      "<NA>\n",
      "DSM 14838\n",
      "not collected\n",
      "not applicable\n",
      "not applicable\n",
      "not applicable\n",
      "not applicable\n",
      "Homo sapiens\n",
      "not collected\n",
      "not collected\n",
      "not collected\n",
      "not collected\n",
      "type strain of Bacteroides cellulosilyticus\n",
      "MIGS/MIMS/MIMARKS.human-gut\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n",
      "<NA>\n"
     ]
    }
   ],
   "source": [
    "for feature in study_df.columns:\n",
    "    print(study_df[feature].iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['run_accession', 'study_accession', 'study_title',\n",
       "       'experiment_accession', 'experiment_title', 'experiment_desc',\n",
       "       'organism_taxid', 'organism_name', 'library_name', 'library_strategy',\n",
       "       'library_source', 'library_selection', 'library_layout',\n",
       "       'sample_accession', 'sample_title', 'biosample', 'bioproject',\n",
       "       'instrument', 'instrument_model', 'instrument_model_desc',\n",
       "       'total_spots', 'total_size', 'run_total_spots', 'run_total_bases',\n",
       "       'run_alias', 'public_filename', 'public_size', 'public_date',\n",
       "       'public_md5', 'public_version', 'public_semantic_name',\n",
       "       'public_supertype', 'public_sratoolkit', 'aws_url', 'aws_free_egress',\n",
       "       'aws_access_type', 'public_url', 'ncbi_url', 'ncbi_free_egress',\n",
       "       'ncbi_access_type', 'gcp_url', 'gcp_free_egress', 'gcp_access_type',\n",
       "       'experiment_alias', 'strain', 'collection_date', 'env_broad_scale',\n",
       "       'env_local_scale', 'env_medium', 'geo_loc_name', 'host',\n",
       "       'isol_growth_condt', 'lat_lon', 'num_replicons', 'ref_biomaterial',\n",
       "       'type-material', 'biosamplemodel', 'sub_species', 'ena_fastq_http',\n",
       "       'ena_fastq_http_1', 'ena_fastq_http_2', 'ena_fastq_ftp',\n",
       "       'ena_fastq_ftp_1', 'ena_fastq_ftp_2'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR24651171 SRR24651172 SRR24651173 SRR24651174 SRR24651175 SRR24651176 SRR24651177 SRR24651178 SRR24651179 SRR24651180 SRR24651181 SRR24651182 SRR24651183 SRR24651184 SRR24651185 SRR24651186 SRR24651187 SRR24651188 SRR24651189 SRR24651190 SRR24651191 SRR24651192 SRR24651193 SRR24651194 SRR24651195 SRR24651196 SRR24651197 SRR24651198 SRR24651199 SRR24651200 SRR24651201 SRR24651202 SRR24651203 SRR24651204 SRR24651205 SRR24651206 SRR24651207 SRR24651208 SRR24651209 SRR24651210 SRR24651211 SRR24651212 SRR24651213 SRR24651214 SRR24651215 SRR24651216 SRR24651217 SRR24651218 SRR24651219 SRR24651220 SRR24651221 SRR24651222 SRR24651223 SRR24651224 SRR24651225 SRR24651226 SRR24651227 SRR24651228 SRR24651229 SRR24651230 SRR24651231 SRR24651232 SRR24651233 SRR24651234 SRR24651235 SRR24651236 SRR24651237 SRR24651238 SRR24651239 SRR24651240 SRR24651241 SRR24651242 SRR24651243 SRR24651244 SRR24651245 SRR24651246 SRR24651247 SRR24651248 SRR24651249 SRR24651250 SRR24651251 SRR24651252 SRR24651253 SRR24651254 SRR24651255 SRR24651256 SRR24651257 SRR24651258 SRR24651259 SRR24651260 SRR24651261 SRR24651262 SRR24651263 SRR24651264 SRR24651265 SRR24651266 SRR24651267 SRR24651268 SRR24651269 SRR24651270 SRR24651271 SRR24651272 SRR24651273 SRR24651274 SRR24651275 SRR24651276 SRR24651277 SRR24651278 SRR24651279 SRR24651280 SRR24651281 SRR24651282 SRR24651283 SRR24651284 SRR24651285 SRR24651286 SRR24651287 SRR24651288 SRR24651289 SRR24651290 SRR24651291 SRR24651292 SRR24651293 SRR24651294 SRR24651295 SRR24651296 SRR24651297 SRR24651298 SRR24651299 SRR24651300 SRR24651301 SRR24651302 SRR24651303 SRR24651304 SRR24651305 SRR24651306 SRR24651307 SRR24651308 SRR24651309 SRR24651310 SRR24651311 SRR24651312 SRR24651313 SRR24651314 SRR24651315 SRR24651316 SRR24651317 SRR24651318 SRR24651319 SRR24651320 SRR24651321 SRR24651322 SRR24651323 SRR24651324 SRR24651325 SRR24651326 SRR24651327 SRR24651328 "
     ]
    }
   ],
   "source": [
    "for accession in study_df[\"run_accession\"]:\n",
    "    print(accession, end=\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.undefined.undefined"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
