{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read sylph results\n",
    "import pandas as pd\n",
    "\n",
    "sylph_results = pd.read_csv(\"/mnt/c/Users/zhenh/tax_data/profile.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_235218/372835136.py:4: DtypeWarning: Columns (61,63,65,74,82,83) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gtdb_metadata = pd.read_csv('/mnt/c/Users/zhenh/tax_data/bac120_metadata.tsv',sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "# Read GTDB metadata\n",
    "import pandas as pd\n",
    "\n",
    "gtdb_metadata = pd.read_csv('/mnt/c/Users/zhenh/tax_data/bac120_metadata.tsv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>ambiguous_bases</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_contamination</th>\n",
       "      <th>checkm_marker_count</th>\n",
       "      <th>checkm_marker_lineage</th>\n",
       "      <th>checkm_marker_set_count</th>\n",
       "      <th>checkm_strain_heterogeneity</th>\n",
       "      <th>coding_bases</th>\n",
       "      <th>coding_density</th>\n",
       "      <th>...</th>\n",
       "      <th>ssu_silva_blast_align_len</th>\n",
       "      <th>ssu_silva_blast_bitscore</th>\n",
       "      <th>ssu_silva_blast_evalue</th>\n",
       "      <th>ssu_silva_blast_perc_identity</th>\n",
       "      <th>ssu_silva_blast_subject_id</th>\n",
       "      <th>ssu_silva_taxonomy</th>\n",
       "      <th>total_gap_length</th>\n",
       "      <th>trna_aa_count</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>trna_selenocysteine_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RS_GCF_001206855.1</td>\n",
       "      <td>14</td>\n",
       "      <td>99.60</td>\n",
       "      <td>0.91</td>\n",
       "      <td>761</td>\n",
       "      <td>g__Streptococcus (UID584)</td>\n",
       "      <td>238</td>\n",
       "      <td>50.00</td>\n",
       "      <td>1827337</td>\n",
       "      <td>86.246727</td>\n",
       "      <td>...</td>\n",
       "      <td>1543</td>\n",
       "      <td>2844</td>\n",
       "      <td>0</td>\n",
       "      <td>99.935</td>\n",
       "      <td>CP028436.1781495.1783046</td>\n",
       "      <td>Bacteria;Firmicutes;Bacilli;Lactobacillales;St...</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RS_GCF_001108865.1</td>\n",
       "      <td>13</td>\n",
       "      <td>99.39</td>\n",
       "      <td>0.34</td>\n",
       "      <td>761</td>\n",
       "      <td>g__Streptococcus (UID584)</td>\n",
       "      <td>238</td>\n",
       "      <td>33.33</td>\n",
       "      <td>1852015</td>\n",
       "      <td>86.277715</td>\n",
       "      <td>...</td>\n",
       "      <td>1543</td>\n",
       "      <td>2844</td>\n",
       "      <td>0</td>\n",
       "      <td>99.935</td>\n",
       "      <td>AKVY01000001.1976120.1977665</td>\n",
       "      <td>Bacteria;Firmicutes;Bacilli;Lactobacillales;St...</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RS_GCF_000730125.1</td>\n",
       "      <td>30</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.92</td>\n",
       "      <td>476</td>\n",
       "      <td>f__Bifidobacteriaceae (UID1462)</td>\n",
       "      <td>217</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2212698</td>\n",
       "      <td>86.271453</td>\n",
       "      <td>...</td>\n",
       "      <td>1522</td>\n",
       "      <td>2811</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>CP008885.1576966.1578495</td>\n",
       "      <td>Bacteria;Actinobacteriota;Actinobacteria;Bifid...</td>\n",
       "      <td>619</td>\n",
       "      <td>19</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RS_GCF_000950275.1</td>\n",
       "      <td>0</td>\n",
       "      <td>99.45</td>\n",
       "      <td>0.53</td>\n",
       "      <td>921</td>\n",
       "      <td>f__Vibrionaceae (UID4865)</td>\n",
       "      <td>366</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4128151</td>\n",
       "      <td>84.994806</td>\n",
       "      <td>...</td>\n",
       "      <td>655</td>\n",
       "      <td>1210</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>KY272023.1.1437</td>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;En...</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RS_GCF_000468955.1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>472</td>\n",
       "      <td>o__Lactobacillales (UID543)</td>\n",
       "      <td>265</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2107260</td>\n",
       "      <td>86.823994</td>\n",
       "      <td>...</td>\n",
       "      <td>1545</td>\n",
       "      <td>2848</td>\n",
       "      <td>0</td>\n",
       "      <td>99.935</td>\n",
       "      <td>CP003157.2105633.2107180</td>\n",
       "      <td>Bacteria;Firmicutes;Bacilli;Lactobacillales;St...</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394927</th>\n",
       "      <td>GB_GCA_934776845.1</td>\n",
       "      <td>0</td>\n",
       "      <td>91.57</td>\n",
       "      <td>0.00</td>\n",
       "      <td>149</td>\n",
       "      <td>k__Bacteria (UID2329)</td>\n",
       "      <td>89</td>\n",
       "      <td>0.00</td>\n",
       "      <td>779596</td>\n",
       "      <td>89.630474</td>\n",
       "      <td>...</td>\n",
       "      <td>1490</td>\n",
       "      <td>1655</td>\n",
       "      <td>0</td>\n",
       "      <td>87.047</td>\n",
       "      <td>FJ951868.1.1481</td>\n",
       "      <td>Bacteria;Firmicutes;Clostridia;Christensenella...</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394928</th>\n",
       "      <td>GB_GCA_023647245.1</td>\n",
       "      <td>4609</td>\n",
       "      <td>99.97</td>\n",
       "      <td>0.13</td>\n",
       "      <td>1162</td>\n",
       "      <td>f__Enterobacteriaceae (UID5121)</td>\n",
       "      <td>336</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4707023</td>\n",
       "      <td>87.866887</td>\n",
       "      <td>...</td>\n",
       "      <td>1315</td>\n",
       "      <td>2396</td>\n",
       "      <td>0</td>\n",
       "      <td>99.316</td>\n",
       "      <td>CP033396.457925.459474</td>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;En...</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394929</th>\n",
       "      <td>GB_GCA_020344015.1</td>\n",
       "      <td>0</td>\n",
       "      <td>93.69</td>\n",
       "      <td>3.30</td>\n",
       "      <td>147</td>\n",
       "      <td>k__Bacteria (UID2495)</td>\n",
       "      <td>91</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4223673</td>\n",
       "      <td>87.882712</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>27200</td>\n",
       "      <td>18</td>\n",
       "      <td>46</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394930</th>\n",
       "      <td>GB_GCA_945889495.1</td>\n",
       "      <td>0</td>\n",
       "      <td>90.94</td>\n",
       "      <td>3.23</td>\n",
       "      <td>247</td>\n",
       "      <td>c__Deltaproteobacteria (UID3216)</td>\n",
       "      <td>155</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4966708</td>\n",
       "      <td>87.941933</td>\n",
       "      <td>...</td>\n",
       "      <td>1538</td>\n",
       "      <td>2719</td>\n",
       "      <td>0</td>\n",
       "      <td>98.57</td>\n",
       "      <td>FPLL01004645.16.1553</td>\n",
       "      <td>Bacteria;RCP2-54;metagenome</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>46</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394931</th>\n",
       "      <td>GB_GCA_934500585.1</td>\n",
       "      <td>0</td>\n",
       "      <td>97.99</td>\n",
       "      <td>0.00</td>\n",
       "      <td>263</td>\n",
       "      <td>o__Clostridiales (UID1212)</td>\n",
       "      <td>149</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2274306</td>\n",
       "      <td>91.972719</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>394932 rows × 110 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 accession  ambiguous_bases  checkm_completeness  \\\n",
       "0       RS_GCF_001206855.1               14                99.60   \n",
       "1       RS_GCF_001108865.1               13                99.39   \n",
       "2       RS_GCF_000730125.1               30               100.00   \n",
       "3       RS_GCF_000950275.1                0                99.45   \n",
       "4       RS_GCF_000468955.1                0               100.00   \n",
       "...                    ...              ...                  ...   \n",
       "394927  GB_GCA_934776845.1                0                91.57   \n",
       "394928  GB_GCA_023647245.1             4609                99.97   \n",
       "394929  GB_GCA_020344015.1                0                93.69   \n",
       "394930  GB_GCA_945889495.1                0                90.94   \n",
       "394931  GB_GCA_934500585.1                0                97.99   \n",
       "\n",
       "        checkm_contamination  checkm_marker_count  \\\n",
       "0                       0.91                  761   \n",
       "1                       0.34                  761   \n",
       "2                       0.92                  476   \n",
       "3                       0.53                  921   \n",
       "4                       0.00                  472   \n",
       "...                      ...                  ...   \n",
       "394927                  0.00                  149   \n",
       "394928                  0.13                 1162   \n",
       "394929                  3.30                  147   \n",
       "394930                  3.23                  247   \n",
       "394931                  0.00                  263   \n",
       "\n",
       "                   checkm_marker_lineage  checkm_marker_set_count  \\\n",
       "0              g__Streptococcus (UID584)                      238   \n",
       "1              g__Streptococcus (UID584)                      238   \n",
       "2        f__Bifidobacteriaceae (UID1462)                      217   \n",
       "3              f__Vibrionaceae (UID4865)                      366   \n",
       "4            o__Lactobacillales (UID543)                      265   \n",
       "...                                  ...                      ...   \n",
       "394927             k__Bacteria (UID2329)                       89   \n",
       "394928   f__Enterobacteriaceae (UID5121)                      336   \n",
       "394929             k__Bacteria (UID2495)                       91   \n",
       "394930  c__Deltaproteobacteria (UID3216)                      155   \n",
       "394931        o__Clostridiales (UID1212)                      149   \n",
       "\n",
       "        checkm_strain_heterogeneity  coding_bases  coding_density  ...  \\\n",
       "0                             50.00       1827337       86.246727  ...   \n",
       "1                             33.33       1852015       86.277715  ...   \n",
       "2                              0.00       2212698       86.271453  ...   \n",
       "3                              0.00       4128151       84.994806  ...   \n",
       "4                              0.00       2107260       86.823994  ...   \n",
       "...                             ...           ...             ...  ...   \n",
       "394927                         0.00        779596       89.630474  ...   \n",
       "394928                         0.00       4707023       87.866887  ...   \n",
       "394929                         0.00       4223673       87.882712  ...   \n",
       "394930                         0.00       4966708       87.941933  ...   \n",
       "394931                         0.00       2274306       91.972719  ...   \n",
       "\n",
       "        ssu_silva_blast_align_len  ssu_silva_blast_bitscore  \\\n",
       "0                            1543                      2844   \n",
       "1                            1543                      2844   \n",
       "2                            1522                      2811   \n",
       "3                             655                      1210   \n",
       "4                            1545                      2848   \n",
       "...                           ...                       ...   \n",
       "394927                       1490                      1655   \n",
       "394928                       1315                      2396   \n",
       "394929                       none                      none   \n",
       "394930                       1538                      2719   \n",
       "394931                       none                      none   \n",
       "\n",
       "        ssu_silva_blast_evalue  ssu_silva_blast_perc_identity  \\\n",
       "0                            0                         99.935   \n",
       "1                            0                         99.935   \n",
       "2                            0                            100   \n",
       "3                            0                            100   \n",
       "4                            0                         99.935   \n",
       "...                        ...                            ...   \n",
       "394927                       0                         87.047   \n",
       "394928                       0                         99.316   \n",
       "394929                    none                           none   \n",
       "394930                       0                          98.57   \n",
       "394931                    none                           none   \n",
       "\n",
       "          ssu_silva_blast_subject_id  \\\n",
       "0           CP028436.1781495.1783046   \n",
       "1       AKVY01000001.1976120.1977665   \n",
       "2           CP008885.1576966.1578495   \n",
       "3                    KY272023.1.1437   \n",
       "4           CP003157.2105633.2107180   \n",
       "...                              ...   \n",
       "394927               FJ951868.1.1481   \n",
       "394928        CP033396.457925.459474   \n",
       "394929                          none   \n",
       "394930          FPLL01004645.16.1553   \n",
       "394931                          none   \n",
       "\n",
       "                                       ssu_silva_taxonomy total_gap_length  \\\n",
       "0       Bacteria;Firmicutes;Bacilli;Lactobacillales;St...               21   \n",
       "1       Bacteria;Firmicutes;Bacilli;Lactobacillales;St...                0   \n",
       "2       Bacteria;Actinobacteriota;Actinobacteria;Bifid...              619   \n",
       "3       Bacteria;Proteobacteria;Gammaproteobacteria;En...                0   \n",
       "4       Bacteria;Firmicutes;Bacilli;Lactobacillales;St...                0   \n",
       "...                                                   ...              ...   \n",
       "394927  Bacteria;Firmicutes;Clostridia;Christensenella...                0   \n",
       "394928  Bacteria;Proteobacteria;Gammaproteobacteria;En...                0   \n",
       "394929                                               none            27200   \n",
       "394930                        Bacteria;RCP2-54;metagenome                0   \n",
       "394931                                               none                0   \n",
       "\n",
       "       trna_aa_count trna_count trna_selenocysteine_count  \n",
       "0                 19         35                         0  \n",
       "1                 19         38                         0  \n",
       "2                 19         55                         0  \n",
       "3                 17         77                         1  \n",
       "4                 19         60                         0  \n",
       "...              ...        ...                       ...  \n",
       "394927            19         41                         0  \n",
       "394928            19         80                         1  \n",
       "394929            18         46                         1  \n",
       "394930            18         46                         1  \n",
       "394931            17         42                         0  \n",
       "\n",
       "[394932 rows x 110 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/70 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 29%|██▊       | 20/70 [00:28<01:09,  1.39s/it]WARNING: FAILURE ( Wed May  8 16:30:44 +08 2024 )\n",
      "nquire -url https://eutils.ncbi.nlm.nih.gov/entrez/eutils/ esummary.fcgi -db nuccore -id JAHAAF010000001.1 -version 2.0 -retmode xml -tool edirect -edirect 14.6 -edirect_os Linux -email zhenhao@GflowZH.\n",
      "<ERROR>Invalid uid JAHAAF010000001.1 at position= 0</ERROR>\n",
      "SECOND ATTEMPT\n",
      "WARNING: FAILURE ( Wed May  8 16:30:47 +08 2024 )\n",
      "nquire -url https://eutils.ncbi.nlm.nih.gov/entrez/eutils/ esummary.fcgi -db nuccore -id JAHAAF010000001.1 -version 2.0 -retmode xml -tool edirect -edirect 14.6 -edirect_os Linux -email zhenhao@GflowZH.\n",
      "<ERROR>Invalid uid JAHAAF010000001.1 at position= 0</ERROR>\n",
      "LAST ATTEMPT\n",
      "ERROR: FAILURE ( Wed May  8 16:30:49 +08 2024 )\n",
      "nquire -url https://eutils.ncbi.nlm.nih.gov/entrez/eutils/ esummary.fcgi -db nuccore -id JAHAAF010000001.1 -version 2.0 -retmode xml -tool edirect -edirect 14.6 -edirect_os Linux -email zhenhao@GflowZH.\n",
      "<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n",
      "<!DOCTYPE eSummaryResult PUBLIC \"-//NLM//DTD esummary nuccore 20201205//EN\" \"https://eutils.ncbi.nlm.nih.gov/eutils/dtd/20201205/esummary_nuccore.dtd\">\n",
      "<eSummaryResult>\n",
      "  <ERROR>Invalid uid JAHAAF010000001.1 at position= 0</ERROR>\n",
      "  <DocumentSummarySet status=\"OK\"/>\n",
      "</eSummaryResult>\n",
      "QUERY FAILURE\n",
      "100%|██████████| 70/70 [01:46<00:00,  1.52s/it]\n"
     ]
    }
   ],
   "source": [
    "# Change contig name into taxid\n",
    "import subprocess\n",
    "from tqdm import tqdm\n",
    "\n",
    "def accession_number_to_taxid(accession):\n",
    "    esummary = subprocess.Popen([\"esummary\", \"-db\", \"nuccore\", \"-id\", accession], stdout=subprocess.PIPE)\n",
    "    output = subprocess.check_output([\"xtract\", \"-pattern\", \"DocumentSummary\", \"-element\", \"TaxId\"], stdin=esummary.stdout)\n",
    "    esummary.wait()\n",
    "    if len(output) == 0:\n",
    "        return None\n",
    "    return int(output)\n",
    "\n",
    "taxid = []\n",
    "for name in tqdm(sylph_results[\"Contig_name\"]):\n",
    "    accession = name.split(\" \")[0]\n",
    "    taxid.append(accession_number_to_taxid(accession))\n",
    "\n",
    "sylph_results[\"tax_id\"] = taxid\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100.0002"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'gtdb_genomes_reps_r220/database/GCF/000/425/005/GCF_000425005.1_genomic.fna.gz'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sylph_results[\"Genome_file\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ete3 import NCBITaxa\n",
    "\n",
    "ncbi = NCBITaxa()\n",
    "\n",
    "\n",
    "def get_level(ID, level='genus'):\n",
    "    \"\"\"\n",
    "    Find the genus of a given species ID.\n",
    "    \"\"\"\n",
    "    if ID is None:\n",
    "        return None\n",
    "        \n",
    "    # Get the lineage of the species\n",
    "    try:\n",
    "        lineage = ncbi.get_lineage(int(ID))\n",
    "        ranks = ncbi.get_rank(lineage)\n",
    "    except:\n",
    "        return None\n",
    "\n",
    "    #print(ranks)\n",
    "    for i in ranks:\n",
    "        if ranks[i] == level:\n",
    "            return str(i)\n",
    "        \n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxid = []\n",
    "genus_taxid = []\n",
    "\n",
    "for i in sylph_results[\"tax_id\"]:\n",
    "    species_taxid.append(get_level(i, 'species'))\n",
    "    genus_taxid.append(get_level(i, 'genus'))\n",
    "\n",
    "sylph_results[\"species_taxid\"] = species_taxid\n",
    "sylph_results[\"genus_taxid\"] = genus_taxid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_file</th>\n",
       "      <th>Genome_file</th>\n",
       "      <th>Taxonomic_abundance</th>\n",
       "      <th>Sequence_abundance</th>\n",
       "      <th>Adjusted_ANI</th>\n",
       "      <th>Eff_cov</th>\n",
       "      <th>ANI_5-95_percentile</th>\n",
       "      <th>Eff_lambda</th>\n",
       "      <th>Lambda_5-95_percentile</th>\n",
       "      <th>Median_cov</th>\n",
       "      <th>Mean_cov_geq1</th>\n",
       "      <th>Containment_ind</th>\n",
       "      <th>Naive_ANI</th>\n",
       "      <th>Contig_name</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>genus_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/000/425/00...</td>\n",
       "      <td>5.0287</td>\n",
       "      <td>3.2719</td>\n",
       "      <td>99.68</td>\n",
       "      <td>3.272</td>\n",
       "      <td>NA-NA</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>NA-NA</td>\n",
       "      <td>3</td>\n",
       "      <td>3.272</td>\n",
       "      <td>1989/2194</td>\n",
       "      <td>99.68</td>\n",
       "      <td>NZ_AUKY01000001.1 Solobacterium moorei DSM 229...</td>\n",
       "      <td>1123263.0</td>\n",
       "      <td>102148</td>\n",
       "      <td>123375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/008/121/49...</td>\n",
       "      <td>2.9827</td>\n",
       "      <td>2.9515</td>\n",
       "      <td>99.04</td>\n",
       "      <td>1.941</td>\n",
       "      <td>98.79-99.18</td>\n",
       "      <td>1.941</td>\n",
       "      <td>1.74-2.60</td>\n",
       "      <td>2</td>\n",
       "      <td>2.693</td>\n",
       "      <td>2206/3444</td>\n",
       "      <td>98.57</td>\n",
       "      <td>NZ_CP043051.1 [Ruminococcus] gnavus ATCC 29149...</td>\n",
       "      <td>411470.0</td>\n",
       "      <td>33038</td>\n",
       "      <td>2316020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/000/381/52...</td>\n",
       "      <td>2.9664</td>\n",
       "      <td>1.7408</td>\n",
       "      <td>99.86</td>\n",
       "      <td>1.930</td>\n",
       "      <td>99.79-100.00</td>\n",
       "      <td>1.930</td>\n",
       "      <td>1.74-2.06</td>\n",
       "      <td>2</td>\n",
       "      <td>2.382</td>\n",
       "      <td>1724/2089</td>\n",
       "      <td>99.38</td>\n",
       "      <td>NZ_KB906605.1 Peptostreptococcus anaerobius VP...</td>\n",
       "      <td>1035196.0</td>\n",
       "      <td>1261</td>\n",
       "      <td>1257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/013/009/55...</td>\n",
       "      <td>2.9076</td>\n",
       "      <td>4.5559</td>\n",
       "      <td>99.32</td>\n",
       "      <td>1.892</td>\n",
       "      <td>99.16-99.42</td>\n",
       "      <td>1.892</td>\n",
       "      <td>1.75-2.24</td>\n",
       "      <td>2</td>\n",
       "      <td>2.583</td>\n",
       "      <td>3655/5274</td>\n",
       "      <td>98.82</td>\n",
       "      <td>NZ_CP046176.1 Phocaeicola dorei strain DSM 178...</td>\n",
       "      <td>357276.0</td>\n",
       "      <td>357276</td>\n",
       "      <td>909656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/001/025/19...</td>\n",
       "      <td>2.7174</td>\n",
       "      <td>1.5740</td>\n",
       "      <td>99.03</td>\n",
       "      <td>1.768</td>\n",
       "      <td>98.91-99.19</td>\n",
       "      <td>1.768</td>\n",
       "      <td>1.52-1.94</td>\n",
       "      <td>2</td>\n",
       "      <td>2.139</td>\n",
       "      <td>1202/1947</td>\n",
       "      <td>98.46</td>\n",
       "      <td>NZ_AP012325.1 Bifidobacterium catenulatum DSM ...</td>\n",
       "      <td>566552.0</td>\n",
       "      <td>1686</td>\n",
       "      <td>1678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/000/157/99...</td>\n",
       "      <td>0.3945</td>\n",
       "      <td>0.4321</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.257</td>\n",
       "      <td>99.64-100.60</td>\n",
       "      <td>0.257</td>\n",
       "      <td>0.21-0.30</td>\n",
       "      <td>1</td>\n",
       "      <td>1.317</td>\n",
       "      <td>856/3688</td>\n",
       "      <td>95.40</td>\n",
       "      <td>NZ_GG657585.1 Holdemania filiformis DSM 12042 ...</td>\n",
       "      <td>545696.0</td>\n",
       "      <td>61171</td>\n",
       "      <td>61170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/025/150/56...</td>\n",
       "      <td>0.3434</td>\n",
       "      <td>0.1880</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.223</td>\n",
       "      <td>99.30-100.95</td>\n",
       "      <td>0.223</td>\n",
       "      <td>0.16-0.29</td>\n",
       "      <td>1</td>\n",
       "      <td>1.113</td>\n",
       "      <td>390/1865</td>\n",
       "      <td>95.08</td>\n",
       "      <td>NZ_CP102283.1 Granulicatella adiacens ATCC 491...</td>\n",
       "      <td>638301.0</td>\n",
       "      <td>46124</td>\n",
       "      <td>117563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/000/154/46...</td>\n",
       "      <td>0.2585</td>\n",
       "      <td>0.1837</td>\n",
       "      <td>99.70</td>\n",
       "      <td>0.168</td>\n",
       "      <td>98.93-100.90</td>\n",
       "      <td>0.168</td>\n",
       "      <td>0.11-0.21</td>\n",
       "      <td>1</td>\n",
       "      <td>1.113</td>\n",
       "      <td>354/2455</td>\n",
       "      <td>93.94</td>\n",
       "      <td>NZ_DS499581.1 Alistipes putredinis DSM 17216 S...</td>\n",
       "      <td>445970.0</td>\n",
       "      <td>28117</td>\n",
       "      <td>239759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/002/222/59...</td>\n",
       "      <td>0.2569</td>\n",
       "      <td>0.2194</td>\n",
       "      <td>99.09</td>\n",
       "      <td>0.167</td>\n",
       "      <td>98.26-100.21</td>\n",
       "      <td>0.167</td>\n",
       "      <td>0.11-0.21</td>\n",
       "      <td>1</td>\n",
       "      <td>1.097</td>\n",
       "      <td>340/2868</td>\n",
       "      <td>93.35</td>\n",
       "      <td>NZ_CP022413.2 Blautia hansenii DSM 20583 chrom...</td>\n",
       "      <td>537007.0</td>\n",
       "      <td>1322</td>\n",
       "      <td>572511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/000/025/30...</td>\n",
       "      <td>0.1144</td>\n",
       "      <td>0.0743</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.074</td>\n",
       "      <td>99.30-102.78</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.03-0.12</td>\n",
       "      <td>1</td>\n",
       "      <td>1.036</td>\n",
       "      <td>195/2083</td>\n",
       "      <td>92.64</td>\n",
       "      <td>NC_013740.1 Acidaminococcus fermentans DSM 207...</td>\n",
       "      <td>591001.0</td>\n",
       "      <td>905</td>\n",
       "      <td>904</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Sample_file                                        Genome_file  \\\n",
       "0   sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/000/425/00...   \n",
       "1   sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/008/121/49...   \n",
       "2   sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/000/381/52...   \n",
       "3   sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/013/009/55...   \n",
       "4   sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/001/025/19...   \n",
       "..                  ...                                                ...   \n",
       "65  sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/000/157/99...   \n",
       "66  sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/025/150/56...   \n",
       "67  sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/000/154/46...   \n",
       "68  sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/002/222/59...   \n",
       "69  sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/000/025/30...   \n",
       "\n",
       "    Taxonomic_abundance  Sequence_abundance  Adjusted_ANI  Eff_cov  \\\n",
       "0                5.0287              3.2719         99.68    3.272   \n",
       "1                2.9827              2.9515         99.04    1.941   \n",
       "2                2.9664              1.7408         99.86    1.930   \n",
       "3                2.9076              4.5559         99.32    1.892   \n",
       "4                2.7174              1.5740         99.03    1.768   \n",
       "..                  ...                 ...           ...      ...   \n",
       "65               0.3945              0.4321        100.00    0.257   \n",
       "66               0.3434              0.1880        100.00    0.223   \n",
       "67               0.2585              0.1837         99.70    0.168   \n",
       "68               0.2569              0.2194         99.09    0.167   \n",
       "69               0.1144              0.0743        100.00    0.074   \n",
       "\n",
       "   ANI_5-95_percentile Eff_lambda Lambda_5-95_percentile  Median_cov  \\\n",
       "0                NA-NA       HIGH                  NA-NA           3   \n",
       "1          98.79-99.18      1.941              1.74-2.60           2   \n",
       "2         99.79-100.00      1.930              1.74-2.06           2   \n",
       "3          99.16-99.42      1.892              1.75-2.24           2   \n",
       "4          98.91-99.19      1.768              1.52-1.94           2   \n",
       "..                 ...        ...                    ...         ...   \n",
       "65        99.64-100.60      0.257              0.21-0.30           1   \n",
       "66        99.30-100.95      0.223              0.16-0.29           1   \n",
       "67        98.93-100.90      0.168              0.11-0.21           1   \n",
       "68        98.26-100.21      0.167              0.11-0.21           1   \n",
       "69        99.30-102.78      0.074              0.03-0.12           1   \n",
       "\n",
       "    Mean_cov_geq1 Containment_ind  Naive_ANI  \\\n",
       "0           3.272       1989/2194      99.68   \n",
       "1           2.693       2206/3444      98.57   \n",
       "2           2.382       1724/2089      99.38   \n",
       "3           2.583       3655/5274      98.82   \n",
       "4           2.139       1202/1947      98.46   \n",
       "..            ...             ...        ...   \n",
       "65          1.317        856/3688      95.40   \n",
       "66          1.113        390/1865      95.08   \n",
       "67          1.113        354/2455      93.94   \n",
       "68          1.097        340/2868      93.35   \n",
       "69          1.036        195/2083      92.64   \n",
       "\n",
       "                                          Contig_name     tax_id  \\\n",
       "0   NZ_AUKY01000001.1 Solobacterium moorei DSM 229...  1123263.0   \n",
       "1   NZ_CP043051.1 [Ruminococcus] gnavus ATCC 29149...   411470.0   \n",
       "2   NZ_KB906605.1 Peptostreptococcus anaerobius VP...  1035196.0   \n",
       "3   NZ_CP046176.1 Phocaeicola dorei strain DSM 178...   357276.0   \n",
       "4   NZ_AP012325.1 Bifidobacterium catenulatum DSM ...   566552.0   \n",
       "..                                                ...        ...   \n",
       "65  NZ_GG657585.1 Holdemania filiformis DSM 12042 ...   545696.0   \n",
       "66  NZ_CP102283.1 Granulicatella adiacens ATCC 491...   638301.0   \n",
       "67  NZ_DS499581.1 Alistipes putredinis DSM 17216 S...   445970.0   \n",
       "68  NZ_CP022413.2 Blautia hansenii DSM 20583 chrom...   537007.0   \n",
       "69  NC_013740.1 Acidaminococcus fermentans DSM 207...   591001.0   \n",
       "\n",
       "   species_taxid genus_taxid  \n",
       "0         102148      123375  \n",
       "1          33038     2316020  \n",
       "2           1261        1257  \n",
       "3         357276      909656  \n",
       "4           1686        1678  \n",
       "..           ...         ...  \n",
       "65         61171       61170  \n",
       "66         46124      117563  \n",
       "67         28117      239759  \n",
       "68          1322      572511  \n",
       "69           905         904  \n",
       "\n",
       "[70 rows x 17 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sylph_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[39], line 11\u001b[0m\n\u001b[1;32m      8\u001b[0m ground_truth_reads \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/mnt/c/Users/zhenh/tax_data/sampled_reads.fastq\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     10\u001b[0m index \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[0;32m---> 11\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m record \u001b[38;5;129;01min\u001b[39;00m SeqIO\u001b[38;5;241m.\u001b[39mparse(ground_truth_reads, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfastq\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m     12\u001b[0m     read_length\u001b[38;5;241m.\u001b[39mappend(\u001b[38;5;28mlen\u001b[39m(record\u001b[38;5;241m.\u001b[39mseq))\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(ground_truth_label, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m gt:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/Bio/SeqIO/Interfaces.py:74\u001b[0m, in \u001b[0;36mSequenceIterator.__next__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     72\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__next__\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m     73\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 74\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mnext\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecords\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     75\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m     76\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshould_close_stream:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/Bio/SeqIO/QualityIO.py:1085\u001b[0m, in \u001b[0;36mFastqPhredIterator.iterate\u001b[0;34m(self, handle)\u001b[0m\n\u001b[1;32m   1075\u001b[0m \u001b[38;5;66;03m# Originally, I used a list expression for each record:\u001b[39;00m\n\u001b[1;32m   1076\u001b[0m \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m   1077\u001b[0m \u001b[38;5;66;03m# qualities = [ord(letter)-SANGER_SCORE_OFFSET for letter in quality_string]\u001b[39;00m\n\u001b[1;32m   1078\u001b[0m \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m   1079\u001b[0m \u001b[38;5;66;03m# Precomputing is faster, perhaps partly by avoiding the subtractions.\u001b[39;00m\n\u001b[1;32m   1080\u001b[0m q_mapping \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m   1081\u001b[0m     \u001b[38;5;28mchr\u001b[39m(letter): letter \u001b[38;5;241m-\u001b[39m SANGER_SCORE_OFFSET\n\u001b[1;32m   1082\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m letter \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(SANGER_SCORE_OFFSET, \u001b[38;5;241m94\u001b[39m \u001b[38;5;241m+\u001b[39m SANGER_SCORE_OFFSET)\n\u001b[1;32m   1083\u001b[0m }\n\u001b[0;32m-> 1085\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m title_line, seq_string, quality_string \u001b[38;5;129;01min\u001b[39;00m FastqGeneralIterator(handle):\n\u001b[1;32m   1086\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m title2ids:\n\u001b[1;32m   1087\u001b[0m         \u001b[38;5;28mid\u001b[39m, name, descr \u001b[38;5;241m=\u001b[39m title2ids(title_line)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/Bio/SeqIO/QualityIO.py:938\u001b[0m, in \u001b[0;36mFastqGeneralIterator\u001b[0;34m(source)\u001b[0m\n\u001b[1;32m    935\u001b[0m seq_string \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    936\u001b[0m \u001b[38;5;66;03m# There will now be one or more sequence lines; keep going until we\u001b[39;00m\n\u001b[1;32m    937\u001b[0m \u001b[38;5;66;03m# find the \"+\" marking the quality line:\u001b[39;00m\n\u001b[0;32m--> 938\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m line \u001b[38;5;129;01min\u001b[39;00m handle:\n\u001b[1;32m    939\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m line[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m+\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    940\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/codecs.py:319\u001b[0m, in \u001b[0;36mBufferedIncrementalDecoder.decode\u001b[0;34m(self, input, final)\u001b[0m\n\u001b[1;32m    314\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_buffer_decode\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;28minput\u001b[39m, errors, final):\n\u001b[1;32m    315\u001b[0m     \u001b[38;5;66;03m# Overwrite this method in subclasses: It must decode input\u001b[39;00m\n\u001b[1;32m    316\u001b[0m     \u001b[38;5;66;03m# and return an (output, length consumed) tuple\u001b[39;00m\n\u001b[1;32m    317\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mNotImplementedError\u001b[39;00m\n\u001b[0;32m--> 319\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdecode\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;28minput\u001b[39m, final\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m):\n\u001b[1;32m    320\u001b[0m     \u001b[38;5;66;03m# decode input (taking the buffer into account)\u001b[39;00m\n\u001b[1;32m    321\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuffer \u001b[38;5;241m+\u001b[39m \u001b[38;5;28minput\u001b[39m\n\u001b[1;32m    322\u001b[0m     (result, consumed) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_buffer_decode(data, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39merrors, final)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Get the ground truth abundance\n",
    "from Bio import SeqIO\n",
    "# Read ground truth file\n",
    "\n",
    "labels = []\n",
    "read_length = []\n",
    "ground_truth_label = \"/mnt/c/Users/zhenh/tax_data/sampled_reads.label\"\n",
    "ground_truth_reads = \"/mnt/c/Users/zhenh/tax_data/sampled_reads.fastq\"\n",
    "\n",
    "index = 0\n",
    "for record in SeqIO.parse(ground_truth_reads, \"fastq\"):\n",
    "    read_length.append(len(record.seq))\n",
    "\n",
    "with open(ground_truth_label, 'r') as gt:\n",
    "    for line in gt:\n",
    "        labels.append(int(line))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ground_truth_df = pd.DataFrame({\"labels\": labels, \"read_length\": read_length})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>labels</th>\n",
       "      <th>read_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100886</td>\n",
       "      <td>8663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100886</td>\n",
       "      <td>2129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100886</td>\n",
       "      <td>3852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100886</td>\n",
       "      <td>6186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100886</td>\n",
       "      <td>790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71995</th>\n",
       "      <td>901</td>\n",
       "      <td>25452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71996</th>\n",
       "      <td>901</td>\n",
       "      <td>2122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71997</th>\n",
       "      <td>901</td>\n",
       "      <td>11199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71998</th>\n",
       "      <td>901</td>\n",
       "      <td>2373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71999</th>\n",
       "      <td>901</td>\n",
       "      <td>2947</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72000 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       labels  read_length\n",
       "0      100886         8663\n",
       "1      100886         2129\n",
       "2      100886         3852\n",
       "3      100886         6186\n",
       "4      100886          790\n",
       "...       ...          ...\n",
       "71995     901        25452\n",
       "71996     901         2122\n",
       "71997     901        11199\n",
       "71998     901         2373\n",
       "71999     901         2947\n",
       "\n",
       "[72000 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ground_truth_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "abundance_df = ground_truth_df.groupby(labels).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "abundance_df[\"read_length\"] = abundance_df[\"read_length\"] / abundance_df[\"read_length\"].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxid = []\n",
    "genus_taxid = []\n",
    "\n",
    "for i in abundance_df.index:\n",
    "    species_taxid.append(get_level(i, 'species'))\n",
    "    genus_taxid.append(get_level(i, 'genus'))\n",
    "\n",
    "abundance_df[\"species_taxid\"] = species_taxid\n",
    "abundance_df[\"genus_taxid\"] = genus_taxid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>labels</th>\n",
       "      <th>read_length</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>genus_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>817000</td>\n",
       "      <td>0.022463</td>\n",
       "      <td>817</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>1636000</td>\n",
       "      <td>0.044365</td>\n",
       "      <td>818</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901</th>\n",
       "      <td>901000</td>\n",
       "      <td>0.015952</td>\n",
       "      <td>901</td>\n",
       "      <td>872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1261</th>\n",
       "      <td>2522000</td>\n",
       "      <td>0.024758</td>\n",
       "      <td>1261</td>\n",
       "      <td>1257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1358</th>\n",
       "      <td>1358000</td>\n",
       "      <td>0.010215</td>\n",
       "      <td>1358</td>\n",
       "      <td>1357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1232452</th>\n",
       "      <td>1232452000</td>\n",
       "      <td>0.020242</td>\n",
       "      <td>1232452</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1232457</th>\n",
       "      <td>1232457000</td>\n",
       "      <td>0.008461</td>\n",
       "      <td>1232457</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297617</th>\n",
       "      <td>1297617000</td>\n",
       "      <td>0.012899</td>\n",
       "      <td>1297617</td>\n",
       "      <td>1392389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1891238</th>\n",
       "      <td>1891238000</td>\n",
       "      <td>0.012329</td>\n",
       "      <td>1891238</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2860054</th>\n",
       "      <td>2860054000</td>\n",
       "      <td>0.009623</td>\n",
       "      <td>2860054</td>\n",
       "      <td>906</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             labels  read_length species_taxid genus_taxid\n",
       "817          817000     0.022463           817         816\n",
       "818         1636000     0.044365           818         816\n",
       "901          901000     0.015952           901         872\n",
       "1261        2522000     0.024758          1261        1257\n",
       "1358        1358000     0.010215          1358        1357\n",
       "...             ...          ...           ...         ...\n",
       "1232452  1232452000     0.020242       1232452        None\n",
       "1232457  1232457000     0.008461       1232457        None\n",
       "1297617  1297617000     0.012899       1297617     1392389\n",
       "1891238  1891238000     0.012329       1891238        None\n",
       "2860054  2860054000     0.009623       2860054         906\n",
       "\n",
       "[70 rows x 4 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abundance_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(sylph_results[\"species_taxid\"]).intersection(abundance_df[\"species_taxid\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_file</th>\n",
       "      <th>Genome_file</th>\n",
       "      <th>Taxonomic_abundance</th>\n",
       "      <th>Sequence_abundance</th>\n",
       "      <th>Adjusted_ANI</th>\n",
       "      <th>Eff_cov</th>\n",
       "      <th>ANI_5-95_percentile</th>\n",
       "      <th>Eff_lambda</th>\n",
       "      <th>Lambda_5-95_percentile</th>\n",
       "      <th>Median_cov</th>\n",
       "      <th>Mean_cov_geq1</th>\n",
       "      <th>Containment_ind</th>\n",
       "      <th>Naive_ANI</th>\n",
       "      <th>Contig_name</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>genus_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/000/210/07...</td>\n",
       "      <td>2.6407</td>\n",
       "      <td>4.3956</td>\n",
       "      <td>99.89</td>\n",
       "      <td>1.718</td>\n",
       "      <td>99.80-99.96</td>\n",
       "      <td>1.718</td>\n",
       "      <td>1.62-1.83</td>\n",
       "      <td>2</td>\n",
       "      <td>2.336</td>\n",
       "      <td>4603/5748</td>\n",
       "      <td>99.29</td>\n",
       "      <td>NC_021017.1 Bacteroides xylanisolvens XB1A, co...</td>\n",
       "      <td>657309.0</td>\n",
       "      <td>371601</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/000/011/06...</td>\n",
       "      <td>2.3081</td>\n",
       "      <td>4.0459</td>\n",
       "      <td>99.64</td>\n",
       "      <td>1.502</td>\n",
       "      <td>99.56-99.73</td>\n",
       "      <td>1.502</td>\n",
       "      <td>1.40-1.58</td>\n",
       "      <td>2</td>\n",
       "      <td>1.993</td>\n",
       "      <td>4166/5940</td>\n",
       "      <td>98.86</td>\n",
       "      <td>NC_004663.1 Bacteroides thetaiotaomicron VPI-5...</td>\n",
       "      <td>226186.0</td>\n",
       "      <td>818</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/028/539/41...</td>\n",
       "      <td>2.0347</td>\n",
       "      <td>3.8416</td>\n",
       "      <td>99.53</td>\n",
       "      <td>1.324</td>\n",
       "      <td>99.44-99.65</td>\n",
       "      <td>1.324</td>\n",
       "      <td>1.24-1.39</td>\n",
       "      <td>2</td>\n",
       "      <td>1.890</td>\n",
       "      <td>4225/6584</td>\n",
       "      <td>98.58</td>\n",
       "      <td>NZ_JANSWE010000001.1 Bacteroides cellulosilyti...</td>\n",
       "      <td>537012.0</td>\n",
       "      <td>246787</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/000/172/17...</td>\n",
       "      <td>1.8388</td>\n",
       "      <td>3.1000</td>\n",
       "      <td>99.99</td>\n",
       "      <td>1.197</td>\n",
       "      <td>99.90-100.12</td>\n",
       "      <td>1.197</td>\n",
       "      <td>1.12-1.25</td>\n",
       "      <td>2</td>\n",
       "      <td>1.895</td>\n",
       "      <td>4084/5808</td>\n",
       "      <td>98.87</td>\n",
       "      <td>NZ_ABJL02000008.1 Bacteroides intestinalis DSM...</td>\n",
       "      <td>471870.0</td>\n",
       "      <td>329854</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/000/025/98...</td>\n",
       "      <td>1.7069</td>\n",
       "      <td>2.4920</td>\n",
       "      <td>98.94</td>\n",
       "      <td>1.111</td>\n",
       "      <td>98.80-99.12</td>\n",
       "      <td>1.111</td>\n",
       "      <td>1.02-1.19</td>\n",
       "      <td>1</td>\n",
       "      <td>1.681</td>\n",
       "      <td>2475/5068</td>\n",
       "      <td>97.71</td>\n",
       "      <td>NC_003228.3 Bacteroides fragilis NCTC 9343, co...</td>\n",
       "      <td>272559.0</td>\n",
       "      <td>817</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCF/000/614/12...</td>\n",
       "      <td>0.9942</td>\n",
       "      <td>1.3527</td>\n",
       "      <td>99.39</td>\n",
       "      <td>0.647</td>\n",
       "      <td>99.15-99.64</td>\n",
       "      <td>0.647</td>\n",
       "      <td>0.57-0.70</td>\n",
       "      <td>1</td>\n",
       "      <td>1.379</td>\n",
       "      <td>1911/4760</td>\n",
       "      <td>97.10</td>\n",
       "      <td>NZ_BAKJ01000148.1 Bacteroides rodentium JCM 16...</td>\n",
       "      <td>1236512.0</td>\n",
       "      <td>691816</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>sampled_reads.fastq</td>\n",
       "      <td>gtdb_genomes_reps_r220/database/GCA/030/295/56...</td>\n",
       "      <td>0.7390</td>\n",
       "      <td>1.0194</td>\n",
       "      <td>99.67</td>\n",
       "      <td>0.481</td>\n",
       "      <td>99.41-99.98</td>\n",
       "      <td>0.481</td>\n",
       "      <td>0.41-0.53</td>\n",
       "      <td>1</td>\n",
       "      <td>1.333</td>\n",
       "      <td>1630/4650</td>\n",
       "      <td>96.67</td>\n",
       "      <td>AP027244.1 Bacteroides finegoldii JCM 13345 DN...</td>\n",
       "      <td>483215.0</td>\n",
       "      <td>338188</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Sample_file                                        Genome_file  \\\n",
       "5   sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/000/210/07...   \n",
       "7   sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/000/011/06...   \n",
       "13  sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/028/539/41...   \n",
       "17  sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/000/172/17...   \n",
       "23  sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/000/025/98...   \n",
       "47  sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCF/000/614/12...   \n",
       "56  sampled_reads.fastq  gtdb_genomes_reps_r220/database/GCA/030/295/56...   \n",
       "\n",
       "    Taxonomic_abundance  Sequence_abundance  Adjusted_ANI  Eff_cov  \\\n",
       "5                2.6407              4.3956         99.89    1.718   \n",
       "7                2.3081              4.0459         99.64    1.502   \n",
       "13               2.0347              3.8416         99.53    1.324   \n",
       "17               1.8388              3.1000         99.99    1.197   \n",
       "23               1.7069              2.4920         98.94    1.111   \n",
       "47               0.9942              1.3527         99.39    0.647   \n",
       "56               0.7390              1.0194         99.67    0.481   \n",
       "\n",
       "   ANI_5-95_percentile Eff_lambda Lambda_5-95_percentile  Median_cov  \\\n",
       "5          99.80-99.96      1.718              1.62-1.83           2   \n",
       "7          99.56-99.73      1.502              1.40-1.58           2   \n",
       "13         99.44-99.65      1.324              1.24-1.39           2   \n",
       "17        99.90-100.12      1.197              1.12-1.25           2   \n",
       "23         98.80-99.12      1.111              1.02-1.19           1   \n",
       "47         99.15-99.64      0.647              0.57-0.70           1   \n",
       "56         99.41-99.98      0.481              0.41-0.53           1   \n",
       "\n",
       "    Mean_cov_geq1 Containment_ind  Naive_ANI  \\\n",
       "5           2.336       4603/5748      99.29   \n",
       "7           1.993       4166/5940      98.86   \n",
       "13          1.890       4225/6584      98.58   \n",
       "17          1.895       4084/5808      98.87   \n",
       "23          1.681       2475/5068      97.71   \n",
       "47          1.379       1911/4760      97.10   \n",
       "56          1.333       1630/4650      96.67   \n",
       "\n",
       "                                          Contig_name     tax_id  \\\n",
       "5   NC_021017.1 Bacteroides xylanisolvens XB1A, co...   657309.0   \n",
       "7   NC_004663.1 Bacteroides thetaiotaomicron VPI-5...   226186.0   \n",
       "13  NZ_JANSWE010000001.1 Bacteroides cellulosilyti...   537012.0   \n",
       "17  NZ_ABJL02000008.1 Bacteroides intestinalis DSM...   471870.0   \n",
       "23  NC_003228.3 Bacteroides fragilis NCTC 9343, co...   272559.0   \n",
       "47  NZ_BAKJ01000148.1 Bacteroides rodentium JCM 16...  1236512.0   \n",
       "56  AP027244.1 Bacteroides finegoldii JCM 13345 DN...   483215.0   \n",
       "\n",
       "   species_taxid genus_taxid  \n",
       "5         371601         816  \n",
       "7            818         816  \n",
       "13        246787         816  \n",
       "17        329854         816  \n",
       "23           817         816  \n",
       "47        691816         816  \n",
       "56        338188         816  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sylph_results[sylph_results[\"genus_taxid\"] == \"816\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>labels</th>\n",
       "      <th>read_length</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>genus_taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>817000</td>\n",
       "      <td>0.022463</td>\n",
       "      <td>817</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>1636000</td>\n",
       "      <td>0.044365</td>\n",
       "      <td>818</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329854</th>\n",
       "      <td>329854000</td>\n",
       "      <td>0.024086</td>\n",
       "      <td>329854</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338188</th>\n",
       "      <td>338188000</td>\n",
       "      <td>0.011071</td>\n",
       "      <td>338188</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371601</th>\n",
       "      <td>371601000</td>\n",
       "      <td>0.031997</td>\n",
       "      <td>371601</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457395</th>\n",
       "      <td>457395000</td>\n",
       "      <td>0.024610</td>\n",
       "      <td>457395</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469588</th>\n",
       "      <td>469588000</td>\n",
       "      <td>0.007296</td>\n",
       "      <td>469588</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469592</th>\n",
       "      <td>469592000</td>\n",
       "      <td>0.012157</td>\n",
       "      <td>469592</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537012</th>\n",
       "      <td>537012000</td>\n",
       "      <td>0.029372</td>\n",
       "      <td>246787</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691816</th>\n",
       "      <td>691816000</td>\n",
       "      <td>0.010310</td>\n",
       "      <td>691816</td>\n",
       "      <td>816</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           labels  read_length species_taxid genus_taxid\n",
       "817        817000     0.022463           817         816\n",
       "818       1636000     0.044365           818         816\n",
       "329854  329854000     0.024086        329854         816\n",
       "338188  338188000     0.011071        338188         816\n",
       "371601  371601000     0.031997        371601         816\n",
       "457395  457395000     0.024610        457395         816\n",
       "469588  469588000     0.007296        469588         816\n",
       "469592  469592000     0.012157        469592         816\n",
       "537012  537012000     0.029372        246787         816\n",
       "691816  691816000     0.010310        691816         816"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abundance_df[abundance_df[\"genus_taxid\"] == \"816\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[62], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mabundance_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43msylph_results\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mon\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mspecies_taxid\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mouter\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlsuffix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m_gt\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrsuffix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m_sylph\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pandas/core/frame.py:9729\u001b[0m, in \u001b[0;36mDataFrame.join\u001b[0;34m(self, other, on, how, lsuffix, rsuffix, sort, validate)\u001b[0m\n\u001b[1;32m   9566\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mjoin\u001b[39m(\n\u001b[1;32m   9567\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   9568\u001b[0m     other: DataFrame \u001b[38;5;241m|\u001b[39m Series \u001b[38;5;241m|\u001b[39m Iterable[DataFrame \u001b[38;5;241m|\u001b[39m Series],\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   9574\u001b[0m     validate: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   9575\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame:\n\u001b[1;32m   9576\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   9577\u001b[0m \u001b[38;5;124;03m    Join columns of another DataFrame.\u001b[39;00m\n\u001b[1;32m   9578\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   9727\u001b[0m \u001b[38;5;124;03m    5  K1  A5   B1\u001b[39;00m\n\u001b[1;32m   9728\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 9729\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_join_compat\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   9730\u001b[0m \u001b[43m        \u001b[49m\u001b[43mother\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9731\u001b[0m \u001b[43m        \u001b[49m\u001b[43mon\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mon\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9732\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhow\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9733\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlsuffix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlsuffix\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9734\u001b[0m \u001b[43m        \u001b[49m\u001b[43mrsuffix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrsuffix\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9735\u001b[0m \u001b[43m        \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9736\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalidate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9737\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pandas/core/frame.py:9768\u001b[0m, in \u001b[0;36mDataFrame._join_compat\u001b[0;34m(self, other, on, how, lsuffix, rsuffix, sort, validate)\u001b[0m\n\u001b[1;32m   9758\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m how \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcross\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m   9759\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m merge(\n\u001b[1;32m   9760\u001b[0m             \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   9761\u001b[0m             other,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   9766\u001b[0m             validate\u001b[38;5;241m=\u001b[39mvalidate,\n\u001b[1;32m   9767\u001b[0m         )\n\u001b[0;32m-> 9768\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmerge\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   9769\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9770\u001b[0m \u001b[43m        \u001b[49m\u001b[43mother\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9771\u001b[0m \u001b[43m        \u001b[49m\u001b[43mleft_on\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mon\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9772\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhow\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9773\u001b[0m \u001b[43m        \u001b[49m\u001b[43mleft_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mon\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m   9774\u001b[0m \u001b[43m        \u001b[49m\u001b[43mright_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m   9775\u001b[0m \u001b[43m        \u001b[49m\u001b[43msuffixes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mlsuffix\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrsuffix\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9776\u001b[0m \u001b[43m        \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9777\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalidate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   9778\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   9779\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   9780\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m on \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pandas/core/reshape/merge.py:148\u001b[0m, in \u001b[0;36mmerge\u001b[0;34m(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[1;32m    131\u001b[0m \u001b[38;5;129m@Substitution\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mleft : DataFrame or named Series\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    132\u001b[0m \u001b[38;5;129m@Appender\u001b[39m(_merge_doc, indents\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m    133\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmerge\u001b[39m(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    146\u001b[0m     validate: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    147\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame:\n\u001b[0;32m--> 148\u001b[0m     op \u001b[38;5;241m=\u001b[39m \u001b[43m_MergeOperation\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    149\u001b[0m \u001b[43m        \u001b[49m\u001b[43mleft\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    150\u001b[0m \u001b[43m        \u001b[49m\u001b[43mright\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    151\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhow\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    152\u001b[0m \u001b[43m        \u001b[49m\u001b[43mon\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mon\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    153\u001b[0m \u001b[43m        \u001b[49m\u001b[43mleft_on\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mleft_on\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    154\u001b[0m \u001b[43m        \u001b[49m\u001b[43mright_on\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mright_on\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    155\u001b[0m \u001b[43m        \u001b[49m\u001b[43mleft_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mleft_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    156\u001b[0m \u001b[43m        \u001b[49m\u001b[43mright_index\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mright_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    157\u001b[0m \u001b[43m        \u001b[49m\u001b[43msort\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    158\u001b[0m \u001b[43m        \u001b[49m\u001b[43msuffixes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msuffixes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    159\u001b[0m \u001b[43m        \u001b[49m\u001b[43mindicator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindicator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    160\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalidate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    161\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    162\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m op\u001b[38;5;241m.\u001b[39mget_result(copy\u001b[38;5;241m=\u001b[39mcopy)\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pandas/core/reshape/merge.py:741\u001b[0m, in \u001b[0;36m_MergeOperation.__init__\u001b[0;34m(self, left, right, how, on, left_on, right_on, axis, left_index, right_index, sort, suffixes, indicator, validate)\u001b[0m\n\u001b[1;32m    733\u001b[0m (\n\u001b[1;32m    734\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mleft_join_keys,\n\u001b[1;32m    735\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mright_join_keys,\n\u001b[1;32m    736\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mjoin_names,\n\u001b[1;32m    737\u001b[0m ) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_merge_keys()\n\u001b[1;32m    739\u001b[0m \u001b[38;5;66;03m# validate the merge keys dtypes. We may need to coerce\u001b[39;00m\n\u001b[1;32m    740\u001b[0m \u001b[38;5;66;03m# to avoid incompatible dtypes\u001b[39;00m\n\u001b[0;32m--> 741\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_maybe_coerce_merge_keys\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    743\u001b[0m \u001b[38;5;66;03m# If argument passed to validate,\u001b[39;00m\n\u001b[1;32m    744\u001b[0m \u001b[38;5;66;03m# check if columns specified as unique\u001b[39;00m\n\u001b[1;32m    745\u001b[0m \u001b[38;5;66;03m# are in fact unique.\u001b[39;00m\n\u001b[1;32m    746\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m validate \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.8/site-packages/pandas/core/reshape/merge.py:1401\u001b[0m, in \u001b[0;36m_MergeOperation._maybe_coerce_merge_keys\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1395\u001b[0m     \u001b[38;5;66;03m# unless we are merging non-string-like with string-like\u001b[39;00m\n\u001b[1;32m   1396\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m (\n\u001b[1;32m   1397\u001b[0m         inferred_left \u001b[38;5;129;01min\u001b[39;00m string_types \u001b[38;5;129;01mand\u001b[39;00m inferred_right \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m string_types\n\u001b[1;32m   1398\u001b[0m     ) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   1399\u001b[0m         inferred_right \u001b[38;5;129;01min\u001b[39;00m string_types \u001b[38;5;129;01mand\u001b[39;00m inferred_left \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m string_types\n\u001b[1;32m   1400\u001b[0m     ):\n\u001b[0;32m-> 1401\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[1;32m   1403\u001b[0m \u001b[38;5;66;03m# datetimelikes must match exactly\u001b[39;00m\n\u001b[1;32m   1404\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m needs_i8_conversion(lk\u001b[38;5;241m.\u001b[39mdtype) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m needs_i8_conversion(rk\u001b[38;5;241m.\u001b[39mdtype):\n",
      "\u001b[0;31mValueError\u001b[0m: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat"
     ]
    }
   ],
   "source": [
    "abundance_df.join(sylph_results, on=\"species_taxid\", how=\"outer\", lsuffix='_gt', rsuffix='_sylph')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{817,\n",
       " 818,\n",
       " 823,\n",
       " 901,\n",
       " 905,\n",
       " 1261,\n",
       " 1264,\n",
       " 1358,\n",
       " 1509,\n",
       " 1535,\n",
       " 1680,\n",
       " 1685,\n",
       " 1686,\n",
       " 1735,\n",
       " 28026,\n",
       " 28117,\n",
       " 28118,\n",
       " 33038,\n",
       " 33039,\n",
       " 35833,\n",
       " 39485,\n",
       " 39488,\n",
       " 40518,\n",
       " 52226,\n",
       " 53443,\n",
       " 61171,\n",
       " 74426,\n",
       " 84030,\n",
       " 84109,\n",
       " 84112,\n",
       " 89153,\n",
       " 100886,\n",
       " 102148,\n",
       " 133926,\n",
       " 154046,\n",
       " 154288,\n",
       " 166486,\n",
       " 208479,\n",
       " 218538,\n",
       " 239935,\n",
       " 246787,\n",
       " 261299,\n",
       " 310297,\n",
       " 310298,\n",
       " 329854,\n",
       " 338188,\n",
       " 357276,\n",
       " 360807,\n",
       " 371601,\n",
       " 387090,\n",
       " 411483,\n",
       " 411486,\n",
       " 691816,\n",
       " 1297617}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(sylph_results[\"species_taxid\"]).intersection(set(abundance_df[\"species_taxid\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(sylph_results[\"genus_taxid\"]).intersection(set(abundance_df[\"genus_taxid\"])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "49"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(sylph_results[\"genus_taxid\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
